// ng-array-view-library/esm5/lib/group/pera-graph/pera-graph-header/pera-graph-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { PeraGraphService } from '../service/pera-graph.service';
import { WarningService } from 'ng-common-library';
import { ArrayViewService } from '../../../service/array-view.service';
var PeraGraphHeaderComponent = /** @class */ (function () {
    function PeraGraphHeaderComponent(peraGraphService, warningService, arrayViewService, ref) {
        this.peraGraphService = peraGraphService;
        this.warningService = warningService;
        this.arrayViewService = arrayViewService;
        this.ref = ref;
    }
    /**
     * @return {?}
     */
    PeraGraphHeaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Subscribe to the chart title subject and invoke change detection when new value is pushed.
        this.arrayViewService.chartTitleSubject$.subscribe(function () {
            if (_this.ref) {
                _this.ref.detectChanges();
            }
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    PeraGraphHeaderComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        // If the latest RI or graph changed, update the header color
        if (changes.latestRi || changes.graph) {
            this.updateHeaderColor();
        }
        this.priorWarningIndicator = this.warningService.getWarningClass(this.priorWarning);
        // This is so that QA can find the prior warning indicator and ensure that it has the correct color
        this.priorWarningIndicatorId = 'priorWarningIndicator-' + this.graph.vnm;
    };
    /**
     * @return {?}
     */
    PeraGraphHeaderComponent.prototype.updateHeaderColor = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var warning = (this.graph) ? this.graph.activeWarning : null;
        this.style = {
            backgroundColor: this.peraGraphService.getBackgroundColorByWarningAndRi(warning, this.latestRi, this.graph.discharged),
            width: '100%',
        };
    };
    PeraGraphHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'phlib-pera-graph-header',
                    template: "<div id=\"{{ arrayViewService.chartTitleSubject$.value.returnFunc(this.graph) }}\" class=\"ph-header\">\n    <!-- Match the background color to the warning level-->\n    <div id=\"headerbackground\" class=\"inner\" [ngStyle]=\"style\">&nbsp;</div>\n    <!-- Display the prior warning indicator on the right side of the header -->\n    <div id=\"{{priorWarningIndicatorId}}}\" class=\"inner {{priorWarningIndicator}}\" style=\"width: 100%; background-color: transparent\"></div>\n    <!-- Display the graph title... could be the Patient's Name, the MRN, the VNM, etc-->\n    <div class=\"inner ph-header-title\" style=\"background-color: transparent\">\n      {{ arrayViewService.chartTitleSubject$.value.returnFunc(this.graph) }}\n    </div>\n    <!-- Display the Rothman Index on the right side of the header -->\n    <div class=\"inner ph-header-score\" style=\"background-color: transparent\">\n        {{ this.peraGraphService.getRiScoreToDisplay(this.latestRi) }}\n    </div>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".inner{position:absolute}.ph-header{color:#fff;width:100%;white-space:nowrap}.ph-header-title{text-shadow:1px 1px 3px #000;font-size:.8em;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:table-cell;vertical-align:middle;padding-left:5px;padding-right:5px;padding-top:1px;width:90%}.ph-header-score{font-size:.8em;text-shadow:1px 1px 3px #000;padding-right:5px;width:100%;text-align:right;align-content:right;padding-top:1px}.hide-on-compact{display:none}div{float:left}#headerbackground{height:16%;border-bottom:1px solid #000}.m-warning{width:0;height:0;padding-left:12%;padding-top:12%;overflow:hidden;float:right}.m-warning:after{content:\"\";display:block;width:0;height:0;margin-left:-500px;margin-top:-500px;float:right;border-left:500px solid transparent;border-top:500px solid #f6cc5d}.h-warning{width:0;height:0;padding-left:12%;padding-top:12%;overflow:hidden;float:right}.h-warning:after{content:\"\";display:block;width:0;height:0;margin-left:-500px;margin-top:-500px;float:right;border-left:500px solid transparent;border-top:500px solid #ee834c}.vh-warning{width:0;height:0;padding-left:12%;padding-top:12%;overflow:hidden;float:right}.vh-warning:after{content:\"\";display:block;width:0;height:0;margin-left:-500px;margin-top:-500px;float:right;border-left:500px solid transparent;border-top:500px solid #d51c19}"]
                }] }
    ];
    /** @nocollapse */
    PeraGraphHeaderComponent.ctorParameters = function () { return [
        { type: PeraGraphService },
        { type: WarningService },
        { type: ArrayViewService },
        { type: ChangeDetectorRef }
    ]; };
    PeraGraphHeaderComponent.propDecorators = {
        graph: [{ type: Input }],
        latestRi: [{ type: Input }],
        priorWarning: [{ type: Input }]
    };
    return PeraGraphHeaderComponent;
}());
export { PeraGraphHeaderComponent };
if (false) {
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.graph;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.latestRi;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.style;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.priorWarningIndicator;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.priorWarningIndicatorId;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.priorWarning;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.peraGraphService;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.warningService;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.arrayViewService;
    /** @type {?} */
    PeraGraphHeaderComponent.prototype.ref;
}
