// ng-array-view-library/esm2015/lib/service/prior-warnings/config/prior-warnings-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PriorWarningsConfigService {
    constructor() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setPriorWarningEndPoint(value) {
        this.priorWarningEndPoint = value;
    }
    /**
     * @return {?}
     */
    getPriorWarningEndPoint() {
        return this.priorWarningEndPoint;
    }
}
PriorWarningsConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PriorWarningsConfigService.ctorParameters = () => [];
/** @nocollapse */ PriorWarningsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function PriorWarningsConfigService_Factory() { return new PriorWarningsConfigService(); }, token: PriorWarningsConfigService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PriorWarningsConfigService.prototype.priorWarningEndPoint;
}
