// ng-common-library/esm5/lib/model/Facility.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var Facility = /** @class */ (function () {
    function Facility(active, facilityID, displayName, displayOrder, facilityTimezone, fullName, primaryType, secondaryType) {
        this.active = active;
        this.facilityID = facilityID;
        this.displayName = displayName;
        this.displayOrder = displayOrder;
        this.facilityTimezone = facilityTimezone;
        this.fullName = fullName;
        this.primaryType = primaryType;
        this.secondaryType = secondaryType;
    }
    return Facility;
}());
export { Facility };
if (false) {
    /** @type {?} */
    Facility.prototype.active;
    /** @type {?} */
    Facility.prototype.facilityID;
    /** @type {?} */
    Facility.prototype.displayName;
    /** @type {?} */
    Facility.prototype.displayOrder;
    /** @type {?} */
    Facility.prototype.facilityTimezone;
    /** @type {?} */
    Facility.prototype.fullName;
    /** @type {?} */
    Facility.prototype.primaryType;
    /** @type {?} */
    Facility.prototype.secondaryType;
}
