// ng-array-view-library/esm5/lib/group/group-header/group-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
var GroupHeaderComponent = /** @class */ (function () {
    function GroupHeaderComponent() {
    }
    /**
     * @return {?}
     */
    GroupHeaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    GroupHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-group-header',
                    template: "<div class=\"group-header\">\n    <div class=\"info\">\n        <div id='{{name}}' class=\"dashboard-grouping-info\">{{name}}</div>\n        <div id='{{name}}-{{numberOfPatients}}' class=\"dashboard-grouping-info pull-right-info\">Patients: {{numberOfPatients}}</div>\n        <hr>\n    </div>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".group-header{margin:.7em 10px 0 3vw;display:block;cursor:pointer}.info{background-color:#f2f2f2;position:relative;z-index:6;height:2.4em}.group-header .dashboard-grouping-info.pull-right-info{font-size:.7em;color:#666;float:right;padding-right:0}.group-header .dashboard-grouping-info{display:inline-block;font-size:1.2em;color:#999;background-color:#f2f2f2;z-index:4;position:relative;padding-right:10px}hr{margin-top:.5em;margin-bottom:.5em;overflow:hidden;border-top:2px solid #898989;position:relative;top:-1.2em;z-index:0}"]
                }] }
    ];
    /** @nocollapse */
    GroupHeaderComponent.ctorParameters = function () { return []; };
    GroupHeaderComponent.propDecorators = {
        numberOfPatients: [{ type: Input }],
        name: [{ type: Input }]
    };
    return GroupHeaderComponent;
}());
export { GroupHeaderComponent };
if (false) {
    /** @type {?} */
    GroupHeaderComponent.prototype.numberOfPatients;
    /** @type {?} */
    GroupHeaderComponent.prototype.name;
}
