// ng-array-view-library/esm5/lib/service/prior-warnings/config/prior-warnings-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var PriorWarningsConfigService = /** @class */ (function () {
    function PriorWarningsConfigService() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    PriorWarningsConfigService.prototype.setPriorWarningEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.priorWarningEndPoint = value;
    };
    /**
     * @return {?}
     */
    PriorWarningsConfigService.prototype.getPriorWarningEndPoint = /**
     * @return {?}
     */
    function () {
        return this.priorWarningEndPoint;
    };
    PriorWarningsConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PriorWarningsConfigService.ctorParameters = function () { return []; };
    /** @nocollapse */ PriorWarningsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function PriorWarningsConfigService_Factory() { return new PriorWarningsConfigService(); }, token: PriorWarningsConfigService, providedIn: "root" });
    return PriorWarningsConfigService;
}());
export { PriorWarningsConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PriorWarningsConfigService.prototype.priorWarningEndPoint;
}
