// ng-login-library/esm5/lib/services/login-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var LoginConfigService = /** @class */ (function () {
    function LoginConfigService() {
    }
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getAppid = /**
     * @return {?}
     */
    function () {
        return this.appid;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setAppid = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.appid = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getTitle = /**
     * @return {?}
     */
    function () {
        return this.title;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setTitle = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.title = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getLoginUrl = /**
     * @return {?}
     */
    function () {
        return this.loginUrl;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setLoginUrl = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.loginUrl = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getLoginApiUrl = /**
     * @return {?}
     */
    function () {
        return this.loginApiUrl;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setLoginApiUrl = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.loginApiUrl = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getLoginIpUrl = /**
     * @return {?}
     */
    function () {
        return this.loginIpUrl;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setLoginIpUrl = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.loginIpUrl = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getLogoutUrl = /**
     * @return {?}
     */
    function () {
        return this.logoutUrl;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setLogoutUrl = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.logoutUrl = value;
    };
    /**
     * @return {?}
     */
    LoginConfigService.prototype.getLogoSrc = /**
     * @return {?}
     */
    function () {
        return this.logoSrc;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LoginConfigService.prototype.setLogoSrc = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.logoSrc = value;
    };
    LoginConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ LoginConfigService.ngInjectableDef = i0.defineInjectable({ factory: function LoginConfigService_Factory() { return new LoginConfigService(); }, token: LoginConfigService, providedIn: "root" });
    return LoginConfigService;
}());
export { LoginConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.title;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginApiUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginIpUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.logoutUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.logoSrc;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.appid;
}
