// ng-login-library/esm5/lib/services/login-ip.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { LoginService } from './login.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./login.service";
var LoginIpService = /** @class */ (function () {
    function LoginIpService(http, loginService) {
        this.http = http;
        this.loginService = loginService;
        this.loginIp$ = new Subject();
        this.loginInfo = null;
    }
    /**
     * @return {?}
     */
    LoginIpService.prototype.getLoginIp = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.http.get(this.loginService.getLoginIpUrl()).subscribe(function (loginIp) {
            console.log('Got login ip: ' + loginIp);
            _this.loginInfo = loginIp;
            _this.loginIp$.next(loginIp);
        }, function (e) {
            console.log('Could not get login ip: ' + e);
            _this.loginInfo = null;
            _this.loginIp$.next(null);
        });
    };
    LoginIpService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LoginIpService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: LoginService }
    ]; };
    /** @nocollapse */ LoginIpService.ngInjectableDef = i0.defineInjectable({ factory: function LoginIpService_Factory() { return new LoginIpService(i0.inject(i1.HttpClient), i0.inject(i2.LoginService)); }, token: LoginIpService, providedIn: "root" });
    return LoginIpService;
}());
export { LoginIpService };
if (false) {
    /** @type {?} */
    LoginIpService.prototype.loginIp$;
    /** @type {?} */
    LoginIpService.prototype.loginInfo;
    /**
     * @type {?}
     * @private
     */
    LoginIpService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    LoginIpService.prototype.loginService;
}
