// ng-filters-library/esm5/lib/filter/diagnosis/diagnosis-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var DiagnosisConfigService = /** @class */ (function () {
    function DiagnosisConfigService() {
        /**
         * class for configuring the endpoint for retrieving all diagnoses
         */
        this.diagnosesEndpoint = null;
    }
    /**
     * @return {?}
     */
    DiagnosisConfigService.prototype.getDiagnosesEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.diagnosesEndpoint) {
            console.warn('Diagnoses endpoint is not set!');
        }
        return this.diagnosesEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DiagnosisConfigService.prototype.setDiagnosesEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.diagnosesEndpoint = value;
    };
    DiagnosisConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ DiagnosisConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisConfigService_Factory() { return new DiagnosisConfigService(); }, token: DiagnosisConfigService, providedIn: "root" });
    return DiagnosisConfigService;
}());
export { DiagnosisConfigService };
if (false) {
    /**
     * class for configuring the endpoint for retrieving all diagnoses
     * @type {?}
     */
    DiagnosisConfigService.prototype.diagnosesEndpoint;
}
