// ng-filters-library/esm2015/lib/filter/diagnosis/diagnosis-rest.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { DiagnosisConfigService } from './diagnosis-config.service';
import { timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./diagnosis-config.service";
export class DiagnosisRestService {
    /**
     * @param {?} http
     * @param {?} diagnosisConfigService
     */
    constructor(http, diagnosisConfigService) {
        this.http = http;
        this.diagnosisConfigService = diagnosisConfigService;
        this.pollInstanceDiagnoses$ = timer(0, 90000);
    }
    /**
     * returns an observable with all diagnoses
     * @return {?}
     */
    getAllDiagnoses() {
        return this.http.get(this.diagnosisConfigService.getDiagnosesEndpoint() + '/all');
    }
    /**
     * returns an observable that broadcasts all active diagnoses on an interval
     * @return {?}
     */
    getAllDiagnosisInterval() {
        return this.pollInstanceDiagnoses$.pipe(switchMap(() => this.getAllDiagnoses()));
    }
}
DiagnosisRestService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DiagnosisRestService.ctorParameters = () => [
    { type: HttpClient },
    { type: DiagnosisConfigService }
];
/** @nocollapse */ DiagnosisRestService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisRestService_Factory() { return new DiagnosisRestService(i0.inject(i1.HttpClient), i0.inject(i2.DiagnosisConfigService)); }, token: DiagnosisRestService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DiagnosisRestService.prototype.pollInstanceDiagnoses$;
    /** @type {?} */
    DiagnosisRestService.prototype.http;
    /** @type {?} */
    DiagnosisRestService.prototype.diagnosisConfigService;
}
