// ng-filters-library/esm2015/lib/filter/diagnosis/diagnosis-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DiagnosisConfigService {
    constructor() {
        /**
         * class for configuring the endpoint for retrieving all diagnoses
         */
        this.diagnosesEndpoint = null;
    }
    /**
     * @return {?}
     */
    getDiagnosesEndpoint() {
        if (!this.diagnosesEndpoint) {
            console.warn('Diagnoses endpoint is not set!');
        }
        return this.diagnosesEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setDiagnosesEndpoint(value) {
        this.diagnosesEndpoint = value;
    }
}
DiagnosisConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ DiagnosisConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisConfigService_Factory() { return new DiagnosisConfigService(); }, token: DiagnosisConfigService, providedIn: "root" });
if (false) {
    /**
     * class for configuring the endpoint for retrieving all diagnoses
     * @type {?}
     */
    DiagnosisConfigService.prototype.diagnosesEndpoint;
}
