// ng-filters-library/esm2015/lib/filter/diagnosis/diagnosis-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DiagnosisFilterService {
    constructor() {
        this.selectedDiagnoses = {};
        this.diagnoses = [];
        this.diagnosesLoaded$ = new EventEmitter();
        this.toggleAll = new EventEmitter();
    }
}
DiagnosisFilterService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DiagnosisFilterService.ctorParameters = () => [];
/** @nocollapse */ DiagnosisFilterService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisFilterService_Factory() { return new DiagnosisFilterService(); }, token: DiagnosisFilterService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DiagnosisFilterService.prototype.selectedDiagnoses;
    /** @type {?} */
    DiagnosisFilterService.prototype.diagnoses;
    /** @type {?} */
    DiagnosisFilterService.prototype.diagnosesLoaded$;
    /** @type {?} */
    DiagnosisFilterService.prototype.toggleAll;
}
