// ng-array-view-library/esm2015/lib/service/prior-warnings/api/prior-warnings-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { PriorWarningsConfigService } from '../config/prior-warnings-config.service';
import { BehaviorSubject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/prior-warnings-config.service";
export class PriorWarningsApiService {
    /**
     * @param {?} http
     * @param {?} priorWarningsConfigService
     */
    constructor(http, priorWarningsConfigService) {
        this.http = http;
        this.priorWarningsConfigService = priorWarningsConfigService;
        this.pollInstance$ = timer(0, 60000);
        this._priorWarningSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    getPriorWarningSubjectValue() {
        return this._priorWarningSubject.value;
    }
    /**
     * @return {?}
     */
    getPriorWarningSubject() {
        return this._priorWarningSubject;
    }
    /**
     * @return {?}
     */
    getPriorWarnings() {
        return this.http.get(this.priorWarningsConfigService.getPriorWarningEndPoint(), { observe: 'response' });
    }
    /**
     * @return {?}
     */
    getPriorWarningInterval() {
        return this.pollInstance$.pipe(switchMap(() => this.getPriorWarnings()));
    }
}
PriorWarningsApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PriorWarningsApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: PriorWarningsConfigService }
];
/** @nocollapse */ PriorWarningsApiService.ngInjectableDef = i0.defineInjectable({ factory: function PriorWarningsApiService_Factory() { return new PriorWarningsApiService(i0.inject(i1.HttpClient), i0.inject(i2.PriorWarningsConfigService)); }, token: PriorWarningsApiService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PriorWarningsApiService.prototype._priorWarningSubject;
    /**
     * @type {?}
     * @private
     */
    PriorWarningsApiService.prototype.pollInstance$;
    /**
     * @type {?}
     * @private
     */
    PriorWarningsApiService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    PriorWarningsApiService.prototype.priorWarningsConfigService;
}
