// ng-login-library/esm2015/lib/login.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LoginComponent } from './login.component';
import { LoginService } from './services/login.service';
import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { LoginConfigService } from './services/login-config.service';
import { NgModule } from '@angular/core';
import { LoginIpService } from './services/login-ip.service';
import { HttpClientModule } from '@angular/common/http';
export class LoginModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: LoginModule,
            providers: [
                HttpClientModule,
                LoginService,
                LoginIpService,
                LoginConfigService
            ]
        };
    }
}
LoginModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    BrowserModule,
                    FormsModule
                ],
                declarations: [LoginComponent],
                exports: [LoginComponent]
            },] }
];
