// ng-common-library/esm5/lib/service/locations/config/locations-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var LocationsConfigService = /** @class */ (function () {
    function LocationsConfigService() {
        this.activeFacilitiesEndpoint = null;
        this.activeUnitsEndpoint = null;
    }
    /**
     * @return {?}
     */
    LocationsConfigService.prototype.getActiveFacilitiesEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.activeFacilitiesEndpoint) {
            console.warn('Active facilities endpoint is not set!');
        }
        return this.activeFacilitiesEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LocationsConfigService.prototype.setActiveFacilitiesEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.activeFacilitiesEndpoint = value;
    };
    /**
     * @return {?}
     */
    LocationsConfigService.prototype.getActiveUnitsEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.activeUnitsEndpoint) {
            console.warn('Active units endpoint is not set!');
        }
        return this.activeUnitsEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LocationsConfigService.prototype.setActiveUnitsEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.activeUnitsEndpoint = value;
    };
    LocationsConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ LocationsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function LocationsConfigService_Factory() { return new LocationsConfigService(); }, token: LocationsConfigService, providedIn: "root" });
    return LocationsConfigService;
}());
export { LocationsConfigService };
if (false) {
    /** @type {?} */
    LocationsConfigService.prototype.activeFacilitiesEndpoint;
    /** @type {?} */
    LocationsConfigService.prototype.activeUnitsEndpoint;
}
