// od-virtualscroll/dist/es5/basic.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function IVirtualScrollOptions() { }
function IVirtualScrollOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IVirtualScrollOptions.prototype.itemWidth;
    /** @type {?} */
    IVirtualScrollOptions.prototype.itemHeight;
    /** @type {?|undefined} */
    IVirtualScrollOptions.prototype.numAdditionalRows;
    /** @type {?|undefined} */
    IVirtualScrollOptions.prototype.numLimitColumns;
}
/**
 * @record
 */
export function IVirtualScrollContainer() { }
function IVirtualScrollContainer_tsickle_Closure_declarations() {
    /** @type {?} */
    IVirtualScrollContainer.prototype.width;
    /** @type {?} */
    IVirtualScrollContainer.prototype.height;
}
/**
 * @record
 */
export function IVirtualScrollMeasurement() { }
function IVirtualScrollMeasurement_tsickle_Closure_declarations() {
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.containerWidth;
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.containerHeight;
    /** @type {?|undefined} */
    IVirtualScrollMeasurement.prototype.itemWidth;
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.itemHeight;
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.numPossibleRows;
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.numPossibleColumns;
    /** @type {?} */
    IVirtualScrollMeasurement.prototype.numPossibleItems;
}
/**
 * @record
 */
export function IVirtualScrollWindow() { }
function IVirtualScrollWindow_tsickle_Closure_declarations() {
    /** @type {?} */
    IVirtualScrollWindow.prototype.dataTimestamp;
    /** @type {?} */
    IVirtualScrollWindow.prototype.containerWidth;
    /** @type {?} */
    IVirtualScrollWindow.prototype.containerHeight;
    /** @type {?|undefined} */
    IVirtualScrollWindow.prototype.itemWidth;
    /** @type {?} */
    IVirtualScrollWindow.prototype.itemHeight;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numVirtualItems;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numVirtualRows;
    /** @type {?} */
    IVirtualScrollWindow.prototype.virtualHeight;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numAdditionalRows;
    /** @type {?} */
    IVirtualScrollWindow.prototype.scrollTop;
    /** @type {?} */
    IVirtualScrollWindow.prototype.scrollPercentage;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numActualRows;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numActualColumns;
    /** @type {?} */
    IVirtualScrollWindow.prototype.actualHeight;
    /** @type {?} */
    IVirtualScrollWindow.prototype.numActualItems;
    /** @type {?} */
    IVirtualScrollWindow.prototype.visibleStartRow;
    /** @type {?} */
    IVirtualScrollWindow.prototype.visibleEndRow;
}
/**
 * @record
 */
export function IVirtualScrollState() { }
function IVirtualScrollState_tsickle_Closure_declarations() {
    /** @type {?} */
    IVirtualScrollState.prototype.measurement;
    /** @type {?} */
    IVirtualScrollState.prototype.scrollWindow;
    /** @type {?} */
    IVirtualScrollState.prototype.rows;
    /** @type {?} */
    IVirtualScrollState.prototype.needsCheck;
}
