// od-virtualscroll/dist/es5/measurement.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} rect
 * @param {?} options
 * @return {?}
 */
export function calcMeasure(rect, options) {
    var /** @type {?} */ numPossibleRows = Math.ceil(rect.height / options.itemHeight);
    var /** @type {?} */ numPossibleColumns = options.itemWidth !== undefined ? Math.floor(rect.width / options.itemWidth) : 0;
    return {
        containerHeight: rect.height,
        containerWidth: rect.width,
        itemHeight: options.itemHeight,
        itemWidth: options.itemWidth,
        numPossibleColumns: numPossibleColumns,
        numPossibleItems: numPossibleRows * numPossibleColumns,
        numPossibleRows: numPossibleRows,
    };
}
var /** @type {?} */ clamp = function (min, max, value) { return Math.min(max, Math.max(min, value)); };
var ɵ0 = clamp;
/**
 * @param {?} scrollTop
 * @param {?} measure
 * @param {?} numItems
 * @param {?} dataTimestamp
 * @param {?} options
 * @return {?}
 */
export function calcScrollWindow(scrollTop, measure, numItems, dataTimestamp, options) {
    var /** @type {?} */ numVirtualItems = numItems;
    var /** @type {?} */ requestedColumns = options.numLimitColumns !== undefined ? options.numLimitColumns : measure.numPossibleColumns;
    var /** @type {?} */ numActualColumns = Math.min(numVirtualItems, requestedColumns);
    var /** @type {?} */ numVirtualRows = Math.ceil(numVirtualItems / Math.max(1, numActualColumns));
    var /** @type {?} */ virtualHeight = numVirtualRows * measure.itemHeight;
    var /** @type {?} */ numAdditionalRows = options.numAdditionalRows !== undefined ? options.numAdditionalRows : 1;
    var /** @type {?} */ requestedRows = measure.numPossibleRows + numAdditionalRows;
    var /** @type {?} */ numActualRows = numActualColumns > 0 ? Math.min(requestedRows, numVirtualRows) : 0;
    var /** @type {?} */ actualHeight = numActualRows * measure.itemHeight;
    var /** @type {?} */ visibleEndRow = numActualColumns > 0 && numActualRows > 0 ? clamp(0, numVirtualRows - 1, Math.floor((scrollTop + actualHeight) / measure.itemHeight) - 1) : -1;
    return {
        actualHeight: actualHeight,
        containerHeight: measure.containerHeight,
        containerWidth: measure.containerWidth,
        dataTimestamp: dataTimestamp,
        itemHeight: measure.itemHeight,
        itemWidth: measure.itemWidth,
        numActualColumns: numActualColumns,
        numActualItems: Math.min(numActualRows * numActualColumns, numVirtualItems),
        numActualRows: numActualRows,
        numAdditionalRows: numAdditionalRows,
        numVirtualItems: numVirtualItems,
        numVirtualRows: numVirtualRows,
        scrollPercentage: clamp(0, 100, scrollTop / (virtualHeight - measure.containerHeight)),
        scrollTop: scrollTop,
        virtualHeight: virtualHeight,
        visibleEndRow: visibleEndRow,
        visibleStartRow: visibleEndRow !== -1 ? Math.max(0, visibleEndRow - numActualRows + 1) : -1
    };
}
/**
 * @param {?} scrollWin
 * @return {?}
 */
export function getMaxIndex(scrollWin) {
    return scrollWin.visibleEndRow * scrollWin.numActualColumns + scrollWin.numActualColumns - 1;
}
export { ɵ0 };
