// od-virtualscroll/dist/es5/virtualScroll.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ComponentFactoryResolver, ContentChild, ElementRef, Input, NgZone, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { animationFrameScheduler as animationScheduler, combineLatest, concat, empty, from, fromEvent, merge, Observable, of, Subject } from 'rxjs';
import { concatMap, debounceTime, distinctUntilChanged, filter, map, pairwise, publish, scan, startWith, withLatestFrom } from 'rxjs/operators';
import { CmdOption, CreateItemCmd, CreateRowCmd, NoopCmd, RemoveItemCmd, RemoveRowCmd, ShiftRowCmd, UpdateItemCmd } from './cmd';
import { forColumnsIn, forColumnsInWithPrev, forRowsIn } from './enumerate';
import { calcMeasure, calcScrollWindow, getMaxIndex } from './measurement';
import { ScrollItem } from './scrollItem';
import { ScrollObservableService } from './service';
import { difference, intersection, isEmpty } from './set';
import { SetScrollTopCmd, UserCmdOption } from './userCmd';
import { VirtualRowComponent } from './virtualRow.component';
var VirtualScrollComponent = /** @class */ (function () {
    function VirtualScrollComponent(_elem, _cdr, _componentFactoryResolver, _obsService, _zone) {
        this._elem = _elem;
        this._cdr = _cdr;
        this._componentFactoryResolver = _componentFactoryResolver;
        this._obsService = _obsService;
        this._zone = _zone;
        this.vsData = empty();
        this.vsOptions = empty();
        this.vsResize = empty();
        this.vsUserCmd = empty();
        this.vsDebounceTime = 0;
        this.vsEqualsFunc = function (prevIndex, curIndex) { return prevIndex === curIndex; };
        this.height = 0;
        this.width = '0px';
        this._rowFactory = this._componentFactoryResolver.resolveComponentFactory(VirtualRowComponent);
        this._subs = [];
    }
    /**
     * @return {?}
     */
    VirtualScrollComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ getContainerRect = function () { return _this._elem.nativeElement.getBoundingClientRect(); };
        var /** @type {?} */ getScrollTop = function () { return _this._elem.nativeElement.scrollTop; };
        var /** @type {?} */ setScrollTop = function (scrollTop) { _this._elem.nativeElement.scrollTop = scrollTop; };
        var /** @type {?} */ initData = [];
        var /** @type {?} */ data$ = /** @type {?} */ (this.vsData.pipe(startWith(initData), publish()));
        var /** @type {?} */ dataMeta$ = data$.pipe(map(function (data) { return [(new Date()).getTime(), data.length]; }));
        var /** @type {?} */ defaultOptions = { itemWidth: 100, itemHeight: 100, numAdditionalRows: 1 };
        var /** @type {?} */ options$ = /** @type {?} */ (this.vsOptions.pipe(startWith(defaultOptions), publish()));
        var /** @type {?} */ rect$ = merge(fromEvent(window, 'resize'), this.vsResize).pipe(debounceTime(this.vsDebounceTime, animationScheduler), map(function () { return getContainerRect(); }), startWith(getContainerRect()), map(function (_a) {
            var width = _a.width, height = _a.height;
            return ({ width: width, height: height });
        }));
        var /** @type {?} */ scroll$ = new Subject();
        this._zone.runOutsideAngular(function () {
            _this._subs.push(fromEvent(_this._elem.nativeElement, 'scroll')
                .pipe(debounceTime(_this.vsDebounceTime, animationScheduler))
                .subscribe(function () {
                _this._zone.run(function () { return scroll$.next(); });
            }));
        });
        var /** @type {?} */ scrollTop$ = scroll$.pipe(map(function () { return getScrollTop(); }), startWith(0));
        var /** @type {?} */ measure$ = /** @type {?} */ (combineLatest(rect$, options$).pipe(map(function (_a) {
            var rect = _a[0], options = _a[1];
            return calcMeasure(rect, options);
        }), publish()));
        var /** @type {?} */ scrollWin$ = /** @type {?} */ (combineLatest(scrollTop$, measure$, dataMeta$, options$).pipe(map(function (_a) {
            var scrollTop = _a[0], measurement = _a[1], _b = _a[2], dataTimestamp = _b[0], dataLength = _b[1], options = _a[3];
            return calcScrollWindow(scrollTop, measurement, dataLength, dataTimestamp, options);
        }), distinctUntilChanged(function (prevWin, curWin) {
            return prevWin.visibleStartRow === curWin.visibleStartRow &&
                prevWin.visibleEndRow === curWin.visibleEndRow &&
                prevWin.numActualColumns === curWin.numActualColumns &&
                prevWin.numVirtualItems === curWin.numVirtualItems &&
                prevWin.dataTimestamp === curWin.dataTimestamp;
        }), publish()));
        var /** @type {?} */ dScrollWin$ = scrollWin$.pipe(pairwise());
        var /** @type {?} */ renderCmd$ = /** @type {?} */ (dScrollWin$.pipe(concatMap(function (_a) {
            var prevWin = _a[0], curWin = _a[1];
            var /** @type {?} */ rowsDiffCmd$ = of(new NoopCmd());
            var /** @type {?} */ rowsUpdateCmd$ = of(new NoopCmd());
            var /** @type {?} */ prevIndexMap = {};
            var /** @type {?} */ curIndexMap = {};
            // abs: prevent iterating when prevWin has -1 -> -1
            forRowsIn(Math.abs(prevWin.visibleStartRow), prevWin.visibleEndRow, prevWin.numActualRows, function (row, index) {
                prevIndexMap[index] = row;
            });
            // abs: prevent iterating when curWin has -1 -> -1
            forRowsIn(Math.abs(curWin.visibleStartRow), curWin.visibleEndRow, curWin.numActualRows, function (row, index) {
                curIndexMap[index] = row;
            });
            var /** @type {?} */ removeRowsMap = difference(prevIndexMap, curIndexMap);
            var /** @type {?} */ createRowsMap = difference(curIndexMap, prevIndexMap);
            if (!isEmpty(removeRowsMap)) {
                var /** @type {?} */ removeRowCmds = [];
                var /** @type {?} */ removeItemCmds_1 = [];
                var _loop_1 = function (key) {
                    var /** @type {?} */ rowIndex = parseInt(key, 10);
                    var /** @type {?} */ row = removeRowsMap[key];
                    removeRowCmds.push(new RemoveRowCmd(row, rowIndex));
                    forColumnsIn(0, prevWin.numActualColumns - 1, row, prevWin.numActualColumns, prevWin.numVirtualItems, function (c, dataIndex) {
                        removeItemCmds_1.push(new RemoveItemCmd(row, rowIndex, c, dataIndex));
                    });
                };
                for (var /** @type {?} */ key in removeRowsMap) {
                    _loop_1(key);
                }
                rowsDiffCmd$ = concat(from(removeItemCmds_1.reverse()), from(removeRowCmds));
            }
            else if (!isEmpty(createRowsMap)) {
                var /** @type {?} */ createRowCmds = [];
                var /** @type {?} */ createItemCmds_1 = [];
                var _loop_2 = function (key) {
                    var /** @type {?} */ rowIndex = parseInt(key, 10);
                    var /** @type {?} */ row = createRowsMap[key];
                    createRowCmds.push(new CreateRowCmd(row, rowIndex, row * curWin.itemHeight));
                    forColumnsIn(0, curWin.numActualColumns - 1, row, curWin.numActualColumns, curWin.numVirtualItems, function (c, dataIndex) {
                        createItemCmds_1.push(new CreateItemCmd(row, rowIndex, c, dataIndex));
                    });
                };
                for (var /** @type {?} */ key in createRowsMap) {
                    _loop_2(key);
                }
                rowsDiffCmd$ = concat(from(createRowCmds), from(createItemCmds_1));
            }
            var /** @type {?} */ existingRows = intersection(prevIndexMap, curIndexMap);
            if (!isEmpty(existingRows)) {
                var /** @type {?} */ shiftRowCmds = [];
                var /** @type {?} */ createItemCmds_2 = [];
                var /** @type {?} */ removeItemCmds_2 = [];
                var /** @type {?} */ updateItemCmds_1 = [];
                var /** @type {?} */ columnDiffCreateItemCmds_1 = [];
                var /** @type {?} */ columnDiffRemoveItemCmds_1 = [];
                var /** @type {?} */ columnsDiffStart = Math.min(prevWin.numActualColumns, curWin.numActualColumns);
                var /** @type {?} */ numColumns = curWin.numActualColumns - prevWin.numActualColumns;
                var _loop_3 = function (key) {
                    var /** @type {?} */ rowIndex = parseInt(key, 10);
                    var /** @type {?} */ prevRow = existingRows[key].left;
                    var /** @type {?} */ row = existingRows[key].right;
                    if (row !== prevRow) {
                        shiftRowCmds.push(new ShiftRowCmd(row, rowIndex, row * curWin.itemHeight));
                    }
                    if (row !== prevRow || numColumns !== 0 || prevWin.numVirtualItems <= getMaxIndex(prevWin) || curWin.numVirtualItems <= getMaxIndex(curWin) || prevWin.dataTimestamp !== curWin.dataTimestamp) {
                        forColumnsInWithPrev(0, columnsDiffStart - 1, row, curWin.numActualColumns, prevRow, prevWin.numActualColumns, function (c, dataIndex, prevDataIndex) {
                            if (dataIndex >= curWin.numVirtualItems && prevDataIndex < prevWin.numVirtualItems) {
                                removeItemCmds_2.push(new RemoveItemCmd(row, rowIndex, c, prevDataIndex));
                            }
                            else if (dataIndex < curWin.numVirtualItems && prevDataIndex >= prevWin.numVirtualItems) {
                                createItemCmds_2.push(new CreateItemCmd(row, rowIndex, c, dataIndex));
                            }
                            else if (dataIndex < curWin.numVirtualItems && prevDataIndex < prevWin.numVirtualItems && !_this.vsEqualsFunc(prevDataIndex, dataIndex)) {
                                updateItemCmds_1.push(new UpdateItemCmd(row, rowIndex, c, dataIndex));
                            }
                        });
                    }
                    if (numColumns > 0) {
                        forColumnsIn(columnsDiffStart, curWin.numActualColumns - 1, row, curWin.numActualColumns, curWin.numVirtualItems, function (c, dataIndex) {
                            columnDiffCreateItemCmds_1.push(new CreateItemCmd(row, rowIndex, c, dataIndex));
                        });
                    }
                    else if (numColumns < 0) {
                        forColumnsIn(columnsDiffStart, prevWin.numActualColumns - 1, prevRow, prevWin.numActualColumns, prevWin.numVirtualItems, function (c, dataIndex) {
                            columnDiffRemoveItemCmds_1.push(new RemoveItemCmd(prevRow, rowIndex, c, dataIndex));
                        });
                    }
                };
                for (var /** @type {?} */ key in existingRows) {
                    _loop_3(key);
                }
                rowsUpdateCmd$ = concat(merge(from(removeItemCmds_2.reverse()), from(createItemCmds_2), from(updateItemCmds_1), from(shiftRowCmds)), merge(from(columnDiffRemoveItemCmds_1.reverse()), from(columnDiffCreateItemCmds_1)));
            }
            return merge(rowsDiffCmd$, rowsUpdateCmd$);
        }), publish()));
        var /** @type {?} */ updateScrollWinFunc$ = scrollWin$.pipe(map(function (scrollWindow) { return function (state) {
            state.scrollWindow = scrollWindow;
            _this._obsService.emitScrollWin([scrollWindow]);
            state.needsCheck = true;
            return state;
        }; }));
        var /** @type {?} */ createRowFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.CreateRow; }), map(function (cmd) { return function (state) {
            var /** @type {?} */ newRow = _this._viewContainer.createComponent(_this._rowFactory);
            newRow.instance.setTransform(cmd.initShift);
            state.rows[cmd.actualIndex] = newRow;
            _this._obsService.emitCreateRow([cmd, newRow]);
            state.needsCheck = false;
            return state;
        }; }));
        var /** @type {?} */ removeRowFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.RemoveRow; }), map(function (cmd) { return function (state) {
            var /** @type {?} */ rowComp = state.rows[cmd.actualIndex];
            rowComp.destroy();
            delete state.rows[cmd.actualIndex];
            _this._obsService.emitRemoveRow([cmd, rowComp]);
            state.needsCheck = false;
            return state;
        }; }));
        var /** @type {?} */ shiftRowFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.ShiftRow; }), map(function (cmd) { return function (state) {
            var /** @type {?} */ shift = /** @type {?} */ (cmd);
            var /** @type {?} */ row = state.rows[shift.actualIndex];
            row.instance.updateRow(shift.virtualIndex);
            row.instance.setTransform(shift.shift);
            _this._obsService.emitShiftRow([shift, row]);
            state.needsCheck = false;
            return state;
        }; }));
        var /** @type {?} */ createItemFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.CreateItem; }), withLatestFrom(data$), map(function (_a) {
            var cmd = _a[0], data = _a[1];
            return function (state) {
                var /** @type {?} */ createItem = /** @type {?} */ (cmd);
                var /** @type {?} */ item = new ScrollItem(data[createItem.dataIndex], createItem.virtualIndex, createItem.columnIndex);
                var /** @type {?} */ viewRef = state.rows[createItem.actualIndex].instance.addItem(_this._templateRef, item);
                _this._obsService.emitCreateItem([createItem, item, viewRef]);
                state.needsCheck = false;
                return state;
            };
        }));
        var /** @type {?} */ updateItemFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.UpdateItem; }), withLatestFrom(data$), map(function (_a) {
            var cmd = _a[0], data = _a[1];
            return function (state) {
                var /** @type {?} */ update = /** @type {?} */ (cmd);
                var /** @type {?} */ item = data[update.dataIndex];
                var /** @type {?} */ viewRef = state.rows[update.actualIndex].instance.updateItem(update.columnIndex, item);
                _this._obsService.emitUpdateItem([update, item, viewRef]);
                state.needsCheck = false;
                return state;
            };
        }));
        var /** @type {?} */ removeItemFunc$ = renderCmd$.pipe(filter(function (cmd) { return cmd.cmdType === CmdOption.RemoveItem; }), map(function (cmd) { return function (state) {
            var /** @type {?} */ comp = state.rows[cmd.actualIndex];
            comp.instance.removeItem(cmd.columnIndex);
            _this._obsService.emitRemoveItem([cmd]);
            state.needsCheck = false;
            return state;
        }; }));
        var /** @type {?} */ userCmd$ = /** @type {?} */ (this.vsUserCmd.pipe(publish()));
        var /** @type {?} */ userSetScrollTop$ = userCmd$.pipe(filter(function (cmd) { return cmd.cmdType === UserCmdOption.SetScrollTop; }));
        var /** @type {?} */ focusRowSetScrollTop$ = userCmd$.pipe(filter(function (cmd) { return cmd.cmdType === UserCmdOption.FocusRow; }), withLatestFrom(scrollWin$), map(function (_a) {
            var cmd = _a[0], scrollWin = _a[1];
            var /** @type {?} */ focusRow = /** @type {?} */ (cmd);
            return new SetScrollTopCmd(focusRow.rowIndex * scrollWin.itemHeight);
        }));
        var /** @type {?} */ focusItemSetScrollTop$ = userCmd$.pipe(filter(function (cmd) { return cmd.cmdType === UserCmdOption.FocusItem; }), withLatestFrom(scrollWin$), map(function (_a) {
            var cmd = _a[0], scrollWin = _a[1];
            var /** @type {?} */ focusItem = /** @type {?} */ (cmd);
            return new SetScrollTopCmd(Math.floor(focusItem.itemIndex / scrollWin.numActualColumns) * scrollWin.itemHeight);
        }));
        var /** @type {?} */ setScrollTopFunc$ = merge(userSetScrollTop$, focusRowSetScrollTop$, focusItemSetScrollTop$)
            .pipe(map(function (cmd) { return function (state) {
            setScrollTop(cmd.value);
            state.needsCheck = false;
            return state;
        }; }));
        var /** @type {?} */ scanFunc = function (state, changeFn) { return changeFn(state); };
        // Update store
        var /** @type {?} */ main$ = merge(createRowFunc$, removeRowFunc$, shiftRowFunc$, createItemFunc$, removeItemFunc$, updateItemFunc$, updateScrollWinFunc$, setScrollTopFunc$)
            .pipe(scan(scanFunc, { measurement: null, scrollWindow: null, rows: {}, needsCheck: false }));
        this._subs.push(main$
            .pipe(filter(function (state) { return state.needsCheck && state.scrollWindow !== null; }))
            .subscribe(function (state) {
            _this.height = state.scrollWindow.virtualHeight;
            if (state.scrollWindow.itemWidth === undefined) {
                _this.width = '100%';
            }
            else {
                _this.width = state.scrollWindow.itemWidth * state.scrollWindow.numActualColumns + "px";
            }
            _this._cdr.markForCheck();
        }));
        // Order is important
        this._subs.push(userCmd$.connect());
        this._subs.push(renderCmd$.connect());
        this._subs.push(scrollWin$.connect());
        this._subs.push(measure$.connect());
        this._subs.push(options$.connect());
        this._subs.push(data$.connect());
    };
    /**
     * @return {?}
     */
    VirtualScrollComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._subs.forEach(function (sub) { return sub.unsubscribe(); });
    };
    VirtualScrollComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'od-virtualscroll',
                    styles: ["\n    :host {\n      display: block;\n      height: 100%;\n      overflow-y: scroll;\n    }\n\n    .od-scroll-container {\n      position: relative;\n      width: 100%;\n    }\n  "],
                    template: "\n    <div class=\"od-scroll-container\" [style.width]=\"width\" [style.height.px]=\"height\">\n      <div #viewRef><div>\n    </div>",
                },] },
    ];
    /** @nocollapse */
    VirtualScrollComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: ComponentFactoryResolver },
        { type: ScrollObservableService },
        { type: NgZone }
    ]; };
    VirtualScrollComponent.propDecorators = {
        _templateRef: [{ type: ContentChild, args: [TemplateRef,] }],
        _viewContainer: [{ type: ViewChild, args: ['viewRef', { read: ViewContainerRef },] }],
        vsData: [{ type: Input }],
        vsOptions: [{ type: Input }],
        vsResize: [{ type: Input }],
        vsUserCmd: [{ type: Input }],
        vsDebounceTime: [{ type: Input }],
        vsEqualsFunc: [{ type: Input }]
    };
    return VirtualScrollComponent;
}());
export { VirtualScrollComponent };
function VirtualScrollComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    VirtualScrollComponent.prototype._templateRef;
    /** @type {?} */
    VirtualScrollComponent.prototype._viewContainer;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsData;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsOptions;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsResize;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsUserCmd;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsDebounceTime;
    /** @type {?} */
    VirtualScrollComponent.prototype.vsEqualsFunc;
    /** @type {?} */
    VirtualScrollComponent.prototype.height;
    /** @type {?} */
    VirtualScrollComponent.prototype.width;
    /** @type {?} */
    VirtualScrollComponent.prototype._rowFactory;
    /** @type {?} */
    VirtualScrollComponent.prototype._subs;
    /** @type {?} */
    VirtualScrollComponent.prototype._elem;
    /** @type {?} */
    VirtualScrollComponent.prototype._cdr;
    /** @type {?} */
    VirtualScrollComponent.prototype._componentFactoryResolver;
    /** @type {?} */
    VirtualScrollComponent.prototype._obsService;
    /** @type {?} */
    VirtualScrollComponent.prototype._zone;
}
