// ng-filters-library/esm5/lib/filter/query-filters/filter-query-rest.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { timer } from 'rxjs';
import { HttpClient, HttpParams } from '@angular/common/http';
import { FilterQueryConfigService } from './filter-query-config.service';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./filter-query-config.service";
var FilterQueryRestService = /** @class */ (function () {
    function FilterQueryRestService(http, filterQueryConfig) {
        this.http = http;
        this.filterQueryConfig = filterQueryConfig;
        this.pollInstance$ = timer(0, 900000);
    }
    /**
     * Rest call for getting filtered visits according to current criteria
     * @param {?} queryParamsGetter
     * @return {?}
     */
    FilterQueryRestService.prototype.getFilteredVisits = /**
     * Rest call for getting filtered visits according to current criteria
     * @param {?} queryParamsGetter
     * @return {?}
     */
    function (queryParamsGetter) {
        /** @type {?} */
        var serverFilterCriteria = queryParamsGetter();
        return this.http.get(this.filterQueryConfig.getFilterQueryEndPoint(), { params: this.getQueryParams(serverFilterCriteria) });
    };
    /**
     * get updated filter criteria on an interval
     * @param {?} queryParamsGetter
     * @return {?}
     */
    FilterQueryRestService.prototype.getFilteredVisitsInterval = /**
     * get updated filter criteria on an interval
     * @param {?} queryParamsGetter
     * @return {?}
     */
    function (queryParamsGetter) {
        var _this = this;
        return this.pollInstance$.pipe(switchMap((function () { return _this.getFilteredVisits(queryParamsGetter); })));
    };
    /**
     * returns the filter criteria in the form of a HttpParams object
     */
    /**
     * returns the filter criteria in the form of a HttpParams object
     * @param {?} serverFilterCriteria
     * @return {?}
     */
    FilterQueryRestService.prototype.getQueryParams = /**
     * returns the filter criteria in the form of a HttpParams object
     * @param {?} serverFilterCriteria
     * @return {?}
     */
    function (serverFilterCriteria) {
        // check if object is empty
        if (Object.keys(serverFilterCriteria).length === 0) {
            return null;
        }
        else {
            /** @type {?} */
            var params = new HttpParams();
            // for every key in the object attach a list of params
            for (var key in serverFilterCriteria) {
                if (serverFilterCriteria.hasOwnProperty(key)) {
                    try {
                        // traverse the list of params for same key and add each one
                        for (var _a = tslib_1.__values(serverFilterCriteria[key]), _b = _a.next(); !_b.done; _b = _a.next()) {
                            var criteria = _b.value;
                            criteria = criteria.replace(/,/g, '|');
                            params = params.append(key, criteria);
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
            }
            return params;
        }
        var e_1, _c;
    };
    FilterQueryRestService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FilterQueryRestService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: FilterQueryConfigService }
    ]; };
    /** @nocollapse */ FilterQueryRestService.ngInjectableDef = i0.defineInjectable({ factory: function FilterQueryRestService_Factory() { return new FilterQueryRestService(i0.inject(i1.HttpClient), i0.inject(i2.FilterQueryConfigService)); }, token: FilterQueryRestService, providedIn: "root" });
    return FilterQueryRestService;
}());
export { FilterQueryRestService };
if (false) {
    /** @type {?} */
    FilterQueryRestService.prototype.pollInstance$;
    /** @type {?} */
    FilterQueryRestService.prototype.http;
    /** @type {?} */
    FilterQueryRestService.prototype.filterQueryConfig;
}
