// ng-filters-library/esm5/lib/filter/query-filters/filter-query-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var FilterQueryConfigService = /** @class */ (function () {
    function FilterQueryConfigService() {
        this.filterQueryEndPoint = null;
    }
    /**
     * @return {?}
     */
    FilterQueryConfigService.prototype.getFilterQueryEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.filterQueryEndPoint) {
            console.warn('Filter query endpoint is not set!');
        }
        return this.filterQueryEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FilterQueryConfigService.prototype.setFilterQueryEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.filterQueryEndPoint = value;
    };
    FilterQueryConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ FilterQueryConfigService.ngInjectableDef = i0.defineInjectable({ factory: function FilterQueryConfigService_Factory() { return new FilterQueryConfigService(); }, token: FilterQueryConfigService, providedIn: "root" });
    return FilterQueryConfigService;
}());
export { FilterQueryConfigService };
if (false) {
    /** @type {?} */
    FilterQueryConfigService.prototype.filterQueryEndPoint;
}
