// ng-common-library/esm2015/lib/service/peragraph/api/peragraph-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { PeragraphConfigService } from '../config/peragraph-config.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/peragraph-config.service";
export class PeragraphApiService {
    /**
     * @param {?} http
     * @param {?} peragraphConfigService
     */
    constructor(http, peragraphConfigService) {
        this.http = http;
        this.peragraphConfigService = peragraphConfigService;
        this.pollInstance$ = timer(0, 60000);
        this._peraGraphSubject = new BehaviorSubject([]);
    }
    /**
     * @return {?}
     */
    getPeraGraphSubject() {
        return this._peraGraphSubject;
    }
    /**
     * @return {?}
     */
    getSinglePeragraph() {
        return this.http.get(this.peragraphConfigService.getSinglePeraGraphEndPoint());
    }
    /**
     * @return {?}
     */
    getAllPeragraphs() {
        return this.http.get(this.peragraphConfigService.getAllPeraGraphsEndPoint());
    }
    /**
     * @return {?}
     */
    getPeraGraphInterval() {
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(() => this.getAllPeragraphs()));
        /*
                 TODO Have a different method with the filtering criteria, using the same subject,
                 and call it when the filter changes. Then we push the change.
        
                 Do the same thing with sorting.
                 */
    }
}
PeragraphApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PeragraphApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: PeragraphConfigService }
];
/** @nocollapse */ PeragraphApiService.ngInjectableDef = i0.defineInjectable({ factory: function PeragraphApiService_Factory() { return new PeragraphApiService(i0.inject(i1.HttpClient), i0.inject(i2.PeragraphConfigService)); }, token: PeragraphApiService, providedIn: "root" });
if (false) {
    /** @type {?} */
    PeragraphApiService.prototype._peraGraphSubject;
    /** @type {?} */
    PeragraphApiService.prototype.pollInstance$;
    /** @type {?} */
    PeragraphApiService.prototype.http;
    /** @type {?} */
    PeragraphApiService.prototype.peragraphConfigService;
}
