// ng-login-library/esm2015/lib/services/login-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class LoginConfigService {
    /**
     * @return {?}
     */
    getAppid() {
        return this.appid;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setAppid(value) {
        this.appid = value;
    }
    /**
     * @return {?}
     */
    getTitle() {
        return this.title;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setTitle(value) {
        this.title = value;
    }
    /**
     * @return {?}
     */
    getLoginUrl() {
        return this.loginUrl;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLoginUrl(value) {
        this.loginUrl = value;
    }
    /**
     * @return {?}
     */
    getLoginApiUrl() {
        return this.loginApiUrl;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLoginApiUrl(value) {
        this.loginApiUrl = value;
    }
    /**
     * @return {?}
     */
    getLoginIpUrl() {
        return this.loginIpUrl;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLoginIpUrl(value) {
        this.loginIpUrl = value;
    }
    /**
     * @return {?}
     */
    getLogoutUrl() {
        return this.logoutUrl;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLogoutUrl(value) {
        this.logoutUrl = value;
    }
    /**
     * @return {?}
     */
    getLogoSrc() {
        return this.logoSrc;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLogoSrc(value) {
        this.logoSrc = value;
    }
}
LoginConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ LoginConfigService.ngInjectableDef = i0.defineInjectable({ factory: function LoginConfigService_Factory() { return new LoginConfigService(); }, token: LoginConfigService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.title;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginApiUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.loginIpUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.logoutUrl;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.logoSrc;
    /**
     * @type {?}
     * @private
     */
    LoginConfigService.prototype.appid;
}
