// ng-array-view-library/esm5/lib/group/pera-graph/tooltip/tooltip.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DateService } from 'ng-common-library';
var TooltipComponent = /** @class */ (function () {
    function TooltipComponent(dateService) {
        this.dateService = dateService;
    }
    /**
     * @return {?}
     */
    TooltipComponent.prototype.getAdmitDate = /**
     * @return {?}
     */
    function () {
        if (this.graph && this.graph.timeZone && this.graph.admitDate) {
            return this.dateService.getDateFormat(this.graph.timeZone, this.graph.admitDate, 'MM/DD/YY HH:mm z');
        }
        return '';
    };
    /**
     * @return {?}
     */
    TooltipComponent.prototype.getDischargeDate = /**
     * @return {?}
     */
    function () {
        if (this.graph && this.graph.timeZone && this.graph.discharged && this.graph.dischargeDate) {
            return this.dateService.getDateFormat(this.graph.timeZone, this.graph.dischargeDate, 'MM/DD/YY HH:mm z');
        }
        return '';
    };
    /**
     * @return {?}
     */
    TooltipComponent.prototype.isDischarged = /**
     * @return {?}
     */
    function () {
        return (this.graph && this.graph.isDischarged);
    };
    TooltipComponent.decorators = [
        { type: Component, args: [{
                    selector: 'phlib-chart-tooltip',
                    template: "<span class=\"patientToolTip\">\n  <b>Patient: </b><span style=\"text-transform: capitalize;\">{{graph?.displayName.toLowerCase()}}</span><br>\n  <b>Facility: </b>{{graph?.facilityId}}<br>\n  <b>Unit Room Bed: </b>{{graph?.unitId}} {{graph?.room}} {{graph?.bed}}<br>\n  <b>DOB: </b>{{graph?.patient?.dob | date:'MM/dd/yy'}}<br>\n  <b>Visit: </b>{{graph?.vnm}}<br>\n  <b>MRN: </b>{{graph?.patient.medicalRecordNumber}}<br>\n  <b>Diagnosis: </b>{{graph?.diagnosis}}<br>\n  <b>Admit Date: </b>{{getAdmitDate()}}<br>\n  <span id=\"dischargeDate\" *ngIf=\"isDischarged()\"><b>Discharge Date: </b>{{getDischargeDate()}}<br></span>\n  <b>Provider: </b>{{ graph?.attendingProvider?.firstName}} {{graph?.attendingProvider?.lastName}}<br>\n  <b>Other Visits with RI Scores: </b>{{graph?.numberVisitsWithValidRI}}<br>\n  <span id=\"dischargeName\"*ngIf=\"isDischarged()\"><b>Discharge Disposition: </b>{{graph?.dischargeDispositionDisplayName}}</span>\n</span>\n",
                    styles: [".patientToolTip{font-size:1em}"]
                }] }
    ];
    /** @nocollapse */
    TooltipComponent.ctorParameters = function () { return [
        { type: DateService }
    ]; };
    TooltipComponent.propDecorators = {
        graph: [{ type: Input }]
    };
    return TooltipComponent;
}());
export { TooltipComponent };
if (false) {
    /** @type {?} */
    TooltipComponent.prototype.graph;
    /**
     * @type {?}
     * @private
     */
    TooltipComponent.prototype.dateService;
}
