// ng-common-library/esm2015/lib/service/diagnosis/api/diagnosis-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { DiagnosisConfigService } from '../config/diagnosis-config.service';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/diagnosis-config.service";
export class DiagnosisApiService {
    /**
     * @param {?} http
     * @param {?} diagnosisConfigService
     */
    constructor(http, diagnosisConfigService) {
        this.http = http;
        this.diagnosisConfigService = diagnosisConfigService;
        this.pollInstance$ = timer(0, 60000);
        this.diagnosesSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    getDiagnosesSubject() {
        return this.diagnosesSubject;
    }
    /**
     * @return {?}
     */
    getDiagnosesSubjectValue() {
        return this.diagnosesSubject.value;
    }
    /**
     * @return {?}
     */
    getAllDiagnoses() {
        return this.http.get(this.diagnosisConfigService.getPrimaryDiagnosesEndpoint(), { observe: 'response' });
    }
    /**
     * @return {?}
     */
    getPrimaryDiagnosesInterval() {
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(() => this.getAllDiagnoses()));
    }
}
DiagnosisApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DiagnosisApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: DiagnosisConfigService }
];
/** @nocollapse */ DiagnosisApiService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisApiService_Factory() { return new DiagnosisApiService(i0.inject(i1.HttpClient), i0.inject(i2.DiagnosisConfigService)); }, token: DiagnosisApiService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DiagnosisApiService.prototype.pollInstance$;
    /** @type {?} */
    DiagnosisApiService.prototype.diagnosesSubject;
    /** @type {?} */
    DiagnosisApiService.prototype.http;
    /** @type {?} */
    DiagnosisApiService.prototype.diagnosisConfigService;
}
