// ng-common-library/esm5/lib/service/flowsheets/config/flowsheets-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var FlowsheetsConfigService = /** @class */ (function () {
    function FlowsheetsConfigService() {
        this.singlePeraGraphFlowSheetEndPoint = null;
        this.allPeragraphFlowSheetsEndPoint = null;
        this.detailedFlowSheetColumnEndpoint = null;
        this.latestRiScoresEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    FlowsheetsConfigService.prototype.setSinglePeraGraphFlowSheetEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.singlePeraGraphFlowSheetEndPoint = value;
    };
    /**
     * @return {?}
     */
    FlowsheetsConfigService.prototype.getSinglePeraGraphFlowSheetEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.singlePeraGraphFlowSheetEndPoint) {
            console.warn('Single PeraGraph flowsheet endpoint is not set!');
        }
        return this.singlePeraGraphFlowSheetEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FlowsheetsConfigService.prototype.setAllPeragraphFlowSheetsEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.allPeragraphFlowSheetsEndPoint = value;
    };
    /**
     * @return {?}
     */
    FlowsheetsConfigService.prototype.getAllPeragraphFlowSheetsEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.allPeragraphFlowSheetsEndPoint) {
            console.warn('All PeraGraph flowsheets endpoint is not set!');
        }
        return this.allPeragraphFlowSheetsEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FlowsheetsConfigService.prototype.setDetailedFlowSheetColumnEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.detailedFlowSheetColumnEndpoint = value;
    };
    /**
     * @return {?}
     */
    FlowsheetsConfigService.prototype.getDetailedFlowSheetColumnEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.detailedFlowSheetColumnEndpoint) {
            console.warn('Detailed flowsheet column endpoint is not set!');
        }
        return this.detailedFlowSheetColumnEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FlowsheetsConfigService.prototype.setLatestRiScoresEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.latestRiScoresEndPoint = value;
    };
    /**
     * @return {?}
     */
    FlowsheetsConfigService.prototype.getLatestRiScoresEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.latestRiScoresEndPoint) {
            console.warn('Latest RI Scores endpoint is not set!');
        }
        return this.latestRiScoresEndPoint;
    };
    FlowsheetsConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ FlowsheetsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function FlowsheetsConfigService_Factory() { return new FlowsheetsConfigService(); }, token: FlowsheetsConfigService, providedIn: "root" });
    return FlowsheetsConfigService;
}());
export { FlowsheetsConfigService };
if (false) {
    /** @type {?} */
    FlowsheetsConfigService.prototype.singlePeraGraphFlowSheetEndPoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.allPeragraphFlowSheetsEndPoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.detailedFlowSheetColumnEndpoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.latestRiScoresEndPoint;
}
