// ng-filters-library/esm2015/lib/filter/providers/providers-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ProvidersConfigService {
    constructor() {
        this.allProvidersEndpoint = null;
    }
    /**
     * @return {?}
     */
    getAllProvidersEndpoint() {
        if (!this.allProvidersEndpoint) {
            console.warn('All Providers Endpoint is not set!');
        }
        return this.allProvidersEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setAllProvidersEndpoint(value) {
        this.allProvidersEndpoint = value;
    }
}
ProvidersConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ ProvidersConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ProvidersConfigService_Factory() { return new ProvidersConfigService(); }, token: ProvidersConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ProvidersConfigService.prototype.allProvidersEndpoint;
}
