// ng-filters-library/esm2015/lib/filter/query-filters/filter-query-rest.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { timer } from 'rxjs';
import { HttpClient, HttpParams } from '@angular/common/http';
import { FilterQueryConfigService } from './filter-query-config.service';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./filter-query-config.service";
export class FilterQueryRestService {
    /**
     * @param {?} http
     * @param {?} filterQueryConfig
     */
    constructor(http, filterQueryConfig) {
        this.http = http;
        this.filterQueryConfig = filterQueryConfig;
        this.pollInstance$ = timer(0, 900000);
    }
    /**
     * Rest call for getting filtered visits according to current criteria
     * @param {?} queryParamsGetter
     * @return {?}
     */
    getFilteredVisits(queryParamsGetter) {
        /** @type {?} */
        const serverFilterCriteria = queryParamsGetter();
        return this.http.get(this.filterQueryConfig.getFilterQueryEndPoint(), { params: this.getQueryParams(serverFilterCriteria) });
    }
    /**
     * get updated filter criteria on an interval
     * @param {?} queryParamsGetter
     * @return {?}
     */
    getFilteredVisitsInterval(queryParamsGetter) {
        return this.pollInstance$.pipe(switchMap((() => this.getFilteredVisits(queryParamsGetter))));
    }
    /**
     * returns the filter criteria in the form of a HttpParams object
     * @param {?} serverFilterCriteria
     * @return {?}
     */
    getQueryParams(serverFilterCriteria) {
        // check if object is empty
        if (Object.keys(serverFilterCriteria).length === 0) {
            return null;
        }
        else {
            /** @type {?} */
            let params = new HttpParams();
            // for every key in the object attach a list of params
            for (const key in serverFilterCriteria) {
                if (serverFilterCriteria.hasOwnProperty(key)) {
                    // traverse the list of params for same key and add each one
                    for (let criteria of serverFilterCriteria[key]) {
                        criteria = criteria.replace(/,/g, '|');
                        params = params.append(key, criteria);
                    }
                }
            }
            return params;
        }
    }
}
FilterQueryRestService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
FilterQueryRestService.ctorParameters = () => [
    { type: HttpClient },
    { type: FilterQueryConfigService }
];
/** @nocollapse */ FilterQueryRestService.ngInjectableDef = i0.defineInjectable({ factory: function FilterQueryRestService_Factory() { return new FilterQueryRestService(i0.inject(i1.HttpClient), i0.inject(i2.FilterQueryConfigService)); }, token: FilterQueryRestService, providedIn: "root" });
if (false) {
    /** @type {?} */
    FilterQueryRestService.prototype.pollInstance$;
    /** @type {?} */
    FilterQueryRestService.prototype.http;
    /** @type {?} */
    FilterQueryRestService.prototype.filterQueryConfig;
}
