// ng-filters-library/esm2015/lib/filter/query-filters/filter-query-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class FilterQueryConfigService {
    constructor() {
        this.filterQueryEndPoint = null;
    }
    /**
     * @return {?}
     */
    getFilterQueryEndPoint() {
        if (!this.filterQueryEndPoint) {
            console.warn('Filter query endpoint is not set!');
        }
        return this.filterQueryEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setFilterQueryEndPoint(value) {
        this.filterQueryEndPoint = value;
    }
}
FilterQueryConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ FilterQueryConfigService.ngInjectableDef = i0.defineInjectable({ factory: function FilterQueryConfigService_Factory() { return new FilterQueryConfigService(); }, token: FilterQueryConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    FilterQueryConfigService.prototype.filterQueryEndPoint;
}
