// ng-common-library/esm5/lib/service/token/config/token-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var TokenConfigService = /** @class */ (function () {
    function TokenConfigService() {
        this.checkTokenEndPoint = null;
        this.refreshTokenEndPoint = null;
        this.getTokenEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TokenConfigService.prototype.setCheckTokenEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.checkTokenEndPoint = value;
    };
    /**
     * @return {?}
     */
    TokenConfigService.prototype.getCheckTokenEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.checkTokenEndPoint) {
            console.warn('Check token endpoint is not set!');
        }
        return this.checkTokenEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TokenConfigService.prototype.setRefreshTokenEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.refreshTokenEndPoint = value;
    };
    /**
     * @return {?}
     */
    TokenConfigService.prototype.getRefreshTokenEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.refreshTokenEndPoint) {
            console.warn('Refresh token endpoint is not set!');
        }
        return this.refreshTokenEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TokenConfigService.prototype.setGetTokenEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.getTokenEndPoint = value;
    };
    /**
     * @return {?}
     */
    TokenConfigService.prototype.getGetTokenEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.getTokenEndPoint) {
            console.warn('Get token endpoint is not set!');
        }
        return this.getTokenEndPoint;
    };
    TokenConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ TokenConfigService.ngInjectableDef = i0.defineInjectable({ factory: function TokenConfigService_Factory() { return new TokenConfigService(); }, token: TokenConfigService, providedIn: "root" });
    return TokenConfigService;
}());
export { TokenConfigService };
if (false) {
    /** @type {?} */
    TokenConfigService.prototype.checkTokenEndPoint;
    /** @type {?} */
    TokenConfigService.prototype.refreshTokenEndPoint;
    /** @type {?} */
    TokenConfigService.prototype.getTokenEndPoint;
}
