// ng-array-view-library/esm5/lib/group/pera-graph/pera-graph-tile/pera-graph-tile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, Input } from '@angular/core';
import { PeraGraphService } from '../service/pera-graph.service';
import { WarningService } from 'ng-common-library';
import { ArrayViewService } from '../../../service/array-view.service';
var PeraGraphTileComponent = /** @class */ (function () {
    function PeraGraphTileComponent(peraGraphService, warningService, arrayViewService, ref) {
        this.peraGraphService = peraGraphService;
        this.warningService = warningService;
        this.arrayViewService = arrayViewService;
        this.ref = ref;
    }
    /**
     * @return {?}
     */
    PeraGraphTileComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Subscribe to the chart title subject and invoke change detection when new value is pushed.
        this.arrayViewService.chartTitleSubject$.subscribe(function () {
            if (_this.ref) {
                _this.ref.detectChanges();
            }
        });
    };
    /**
     * @return {?}
     */
    PeraGraphTileComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.updateBackgroundColor();
    };
    /**
     * @return {?}
     */
    PeraGraphTileComponent.prototype.updateBackgroundColor = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var warning = (this.graph) ? this.graph.activeWarning : null;
        this.style = {
            backgroundColor: this.peraGraphService.getBackgroundColorByWarningAndRi(warning, this.latestRi, this.graph.discharged)
        };
        this.priorWarningIndicator = this.warningService.getWarningClass(this.priorWarning);
    };
    PeraGraphTileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'phlib-pera-graph-tile',
                    template: "<!-- Outer div establishes QA test lookup ID, other divs display over the top of each other -->\n<div id=\"{{ arrayViewService.chartTitleSubject$.value.returnFunc(this.graph) }}\" class=\"ph-tile\">\n    <div class=\"inner ph-tile {{ priorWarningIndicator }}\" [ngStyle]=\"style\">\n    </div>\n    <div class=\"inner ph-tile-score\" style=\"background-color: transparent\">\n        {{ this.peraGraphService.getRiScoreToDisplay(this.latestRi) }}\n    </div>\n</div>\n",
                    styles: [".inner{position:absolute}.ph-tile{width:100%;height:100%;text-align:center;cursor:pointer}.ph-tile-score{font-size:1.3em;padding-top:25%;text-align:center;text-shadow:1px 1px 3px #3b3b3b;color:#fff;width:100%;height:100%}.vh-warning{width:0;height:0;border-style:solid;border-width:0 63px 63px 0;border-color:transparent #d51c19 transparent transparent}.h-warning{width:0;height:0;border-style:solid;border-width:0 63px 63px 0;border-color:transparent #fc7134 transparent transparent}.m-warning{width:0;height:0;border-style:solid;border-width:0 63px 63px 0;border-color:transparent #fdc43a transparent transparent}"]
                }] }
    ];
    /** @nocollapse */
    PeraGraphTileComponent.ctorParameters = function () { return [
        { type: PeraGraphService },
        { type: WarningService },
        { type: ArrayViewService },
        { type: ChangeDetectorRef }
    ]; };
    PeraGraphTileComponent.propDecorators = {
        graph: [{ type: Input }],
        latestRi: [{ type: Input }],
        priorWarning: [{ type: Input }]
    };
    return PeraGraphTileComponent;
}());
export { PeraGraphTileComponent };
if (false) {
    /** @type {?} */
    PeraGraphTileComponent.prototype.graph;
    /** @type {?} */
    PeraGraphTileComponent.prototype.latestRi;
    /** @type {?} */
    PeraGraphTileComponent.prototype.priorWarning;
    /** @type {?} */
    PeraGraphTileComponent.prototype.priorWarningIndicator;
    /** @type {?} */
    PeraGraphTileComponent.prototype.style;
    /** @type {?} */
    PeraGraphTileComponent.prototype.peraGraphService;
    /** @type {?} */
    PeraGraphTileComponent.prototype.warningService;
    /** @type {?} */
    PeraGraphTileComponent.prototype.arrayViewService;
    /** @type {?} */
    PeraGraphTileComponent.prototype.ref;
}
