// ng-common-library/esm2015/lib/model/PatientAccess.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export class PatientAccess {
    /**
     * @param {?} viewedBy
     * @param {?} patient
     * @param {?} vipIndicator
     * @param {?} vnm
     * @param {?} mrn
     * @param {?} admitDate
     * @param {?} dischargeDate
     * @param {?} facilityId
     * @param {?} unitId
     * @param {?} provider
     * @param {?} utcCode
     * @param {?} usageCode
     * @param {?} viewDate
     */
    constructor(viewedBy, patient, vipIndicator, vnm, mrn, admitDate, dischargeDate, facilityId, unitId, provider, utcCode, usageCode, viewDate) {
        this.viewedBy = viewedBy;
        this.patient = patient;
        this.vipIndicator = vipIndicator;
        this.vnm = vnm;
        this.mrn = mrn;
        this.admitDate = admitDate;
        this.dischargeDate = dischargeDate;
        this.facilityId = facilityId;
        this.unitId = unitId;
        this.provider = provider;
        this.utcCode = utcCode;
        this.usageCode = usageCode;
        this.viewDate = viewDate;
    }
}
if (false) {
    /** @type {?} */
    PatientAccess.prototype.viewedBy;
    /** @type {?} */
    PatientAccess.prototype.patient;
    /** @type {?} */
    PatientAccess.prototype.vipIndicator;
    /** @type {?} */
    PatientAccess.prototype.vnm;
    /** @type {?} */
    PatientAccess.prototype.mrn;
    /** @type {?} */
    PatientAccess.prototype.admitDate;
    /** @type {?} */
    PatientAccess.prototype.dischargeDate;
    /** @type {?} */
    PatientAccess.prototype.facilityId;
    /** @type {?} */
    PatientAccess.prototype.unitId;
    /** @type {?} */
    PatientAccess.prototype.provider;
    /** @type {?} */
    PatientAccess.prototype.utcCode;
    /** @type {?} */
    PatientAccess.prototype.usageCode;
    /** @type {?} */
    PatientAccess.prototype.viewDate;
}
