// ng-filters-library/esm2015/lib/filter/service-line/service-line-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ServiceLineConfigService {
    constructor() {
        this.serviceLineDescriptionsEndpoint = null;
    }
    /**
     * @return {?}
     */
    getServiceLineDescriptionsEndpoint() {
        if (!this.serviceLineDescriptionsEndpoint) {
            console.warn('Service Line Descriptions endpoint is not set!');
        }
        return this.serviceLineDescriptionsEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setServiceLineDescriptionsEndpoint(value) {
        this.serviceLineDescriptionsEndpoint = value;
    }
}
ServiceLineConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ ServiceLineConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ServiceLineConfigService_Factory() { return new ServiceLineConfigService(); }, token: ServiceLineConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ServiceLineConfigService.prototype.serviceLineDescriptionsEndpoint;
}
