// ng-array-view-library/esm2015/lib/array-view.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ArrayViewComponent } from './array-view.component';
import { GroupComponent } from './group/group.component';
import { PeraGraphComponent } from './group/pera-graph/pera-graph.component';
import { ChartComponent } from './group/pera-graph/chart/chart.component';
import { HighchartsChartModule } from 'highcharts-angular';
import { VirtualScrollModule } from 'od-virtualscroll';
import { ResizableModule } from 'angular-resizable-element';
import { PeraGraphTileComponent } from './group/pera-graph/pera-graph-tile/pera-graph-tile.component';
import { GroupHeaderComponent } from './group/group-header/group-header.component';
import { PeraGraphHeaderComponent } from './group/pera-graph/pera-graph-header/pera-graph-header.component';
import { TooltipComponent } from './group/pera-graph/tooltip/tooltip.component';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { FilterQueryConfigService, FilterQueryRestService, FilterService } from 'ng-filters-library';
export class ArrayViewModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: ArrayViewModule,
            providers: [FilterService, FilterQueryRestService, FilterQueryConfigService]
        };
    }
}
ArrayViewModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule,
                    HighchartsChartModule,
                    VirtualScrollModule,
                    ResizableModule,
                    NgbModule
                ],
                declarations: [
                    ArrayViewComponent,
                    GroupComponent,
                    GroupHeaderComponent,
                    PeraGraphComponent,
                    ChartComponent,
                    PeraGraphTileComponent,
                    TooltipComponent,
                    PeraGraphHeaderComponent
                ],
                exports: [ArrayViewComponent]
            },] }
];
