// ng-array-view-library/esm2015/lib/group/pera-graph/service/pera-graph.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WarningColors } from '../../../model/WarningColors';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class PeraGraphService {
    constructor() {
        // Subject for navigating to the PSG
        this.graphClicked$ = new Subject();
    }
    /**
     * @param {?} graph
     * @return {?}
     */
    graphClickedEvent(graph) {
        this.closeToolTip();
        this.graphClicked$.next(graph);
    }
    /**
     * @param {?} warning
     * @param {?} latestRi
     * @param {?} isDischarged
     * @return {?}
     */
    getBackgroundColorByWarningAndRi(warning, latestRi, isDischarged) {
        /** @type {?} */
        let backgroundColor;
        if (warning && !isDischarged) {
            switch (warning.category) {
                case 'V':
                    backgroundColor = WarningColors.VeryHigh;
                    break;
                case 'H':
                    backgroundColor = WarningColors.High;
                    break;
                case 'M':
                    backgroundColor = WarningColors.Medium;
                    break;
            }
        }
        else if (latestRi && !isDischarged) {
            backgroundColor = WarningColors.NoCurrent;
        }
        else {
            backgroundColor = WarningColors.NoRi;
        }
        return backgroundColor;
    }
    /**
     * @param {?} latestRi
     * @return {?}
     */
    getRiScoreToDisplay(latestRi) {
        return (latestRi) ? latestRi + '' : '--';
    }
    /**
     * @return {?}
     */
    closeToolTip() {
        if (this.tooltip && this.tooltip.isOpen) {
            this.tooltip.close();
        }
    }
}
PeraGraphService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PeraGraphService.ctorParameters = () => [];
/** @nocollapse */ PeraGraphService.ngInjectableDef = i0.defineInjectable({ factory: function PeraGraphService_Factory() { return new PeraGraphService(); }, token: PeraGraphService, providedIn: "root" });
if (false) {
    /** @type {?} */
    PeraGraphService.prototype.graphClicked$;
    /** @type {?} */
    PeraGraphService.prototype.tooltip;
}
