// ng-filters-library/esm2015/lib/group/facility-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LocationSelectorService } from '../filter/locations/location-selector.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../filter/locations/location-selector.service";
export class FacilityGroupService {
    /**
     * @param {?} locationSelectorService
     */
    constructor(locationSelectorService) {
        this.locationSelectorService = locationSelectorService;
        /**
         * things that need to emit before generating the facility lanes
         */
        this.dependency = new Subject();
        this.locationSelectorService.loadedLocations.subscribe(() => {
            this.dependency.next();
        });
    }
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    generateLanes() {
        /** @type {?} */
        const option = {
            groupName: 'Facility',
            lanes: []
        };
        for (const facility of this.locationSelectorService.facilities) {
            option.lanes.push({ laneFunction: this.generateFacilitiyLaneFilterFunction(facility), laneTitle: facility.displayName });
        }
        return option;
    }
    /**
     * generate filter function for grouping by facility
     * @param {?} facility - facility to filter for
     * @return {?}
     */
    generateFacilitiyLaneFilterFunction(facility) {
        /** @type {?} */
        const facilityName = facility.displayName;
        return function (peraGraph) {
            return peraGraph.facilityId === facilityName;
        };
    }
}
FacilityGroupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
FacilityGroupService.ctorParameters = () => [
    { type: LocationSelectorService }
];
/** @nocollapse */ FacilityGroupService.ngInjectableDef = i0.defineInjectable({ factory: function FacilityGroupService_Factory() { return new FacilityGroupService(i0.inject(i1.LocationSelectorService)); }, token: FacilityGroupService, providedIn: "root" });
if (false) {
    /**
     * things that need to emit before generating the facility lanes
     * @type {?}
     */
    FacilityGroupService.prototype.dependency;
    /** @type {?} */
    FacilityGroupService.prototype.locationSelectorService;
}
