// ng-filters-library/esm2015/lib/group/hospital-los-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class HospitalLosGroupService {
    constructor() {
        /**
         * size of LOS lane ranges
         */
        this.losIncrement = 1;
        /**
         * number of lanes LOS group by can have
         */
        this.losLaneMax = 5;
        /**
         * things that need to emit before generating the Hospital Los lanes
         */
        this.dependency = null;
    }
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    generateLanes() {
        /** @type {?} */
        const option = {
            groupName: 'Hospital LOS',
            lanes: []
        };
        for (let i = 0; i < this.losLaneMax * this.losIncrement; i += this.losIncrement) {
            /** @type {?} */
            const losLaneInfo = this.generateLosLaneFilterFunctionAndTitle(i, i + this.losIncrement);
            option.lanes.push({ laneFunction: losLaneInfo.filterFunction, laneTitle: losLaneInfo.title });
        }
        /** @type {?} */
        const losLaneInfoFinal = this.generateLosLaneFilterFunctionAndTitle(this.losLaneMax * this.losIncrement);
        option.lanes.push({ laneFunction: losLaneInfoFinal.filterFunction, laneTitle: losLaneInfoFinal.title });
        return option;
    }
    /**
     * generate filter function and label for Hospital LOS group by
     * @param {?} minIndex - number of days before current time to start including graphs in this lane (inclusive)
     * @param {?=} maxIndex - number of days before current time to stop including graphs (exclusive)
     * @return {?}
     */
    generateLosLaneFilterFunctionAndTitle(minIndex, maxIndex = -1) {
        /** @type {?} */
        const minVolume = minIndex * 86400000;
        if (maxIndex !== -1) {
            /** @type {?} */
            const maxVolume = maxIndex * 86400000;
            return {
                filterFunction: function (peraGraph) {
                    /** @type {?} */
                    const current = ((new Date()).getTime());
                    /** @type {?} */
                    const min = current - minVolume;
                    /** @type {?} */
                    const max = current - maxVolume;
                    return peraGraph.admitDate <= min && peraGraph.admitDate > max;
                }, title: minIndex.toString() + ' - ' + maxIndex.toString() + ' days'
            };
        }
        else {
            // create catch all if max is not included in function call
            return {
                filterFunction: function (peraGraph) {
                    /** @type {?} */
                    const current = ((new Date()).getTime());
                    /** @type {?} */
                    const min = current - minVolume;
                    return peraGraph.admitDate <= min;
                }, title: minIndex.toString() + '+ days'
            };
        }
    }
}
HospitalLosGroupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
HospitalLosGroupService.ctorParameters = () => [];
/** @nocollapse */ HospitalLosGroupService.ngInjectableDef = i0.defineInjectable({ factory: function HospitalLosGroupService_Factory() { return new HospitalLosGroupService(); }, token: HospitalLosGroupService, providedIn: "root" });
if (false) {
    /**
     * size of LOS lane ranges
     * @type {?}
     */
    HospitalLosGroupService.prototype.losIncrement;
    /**
     * number of lanes LOS group by can have
     * @type {?}
     */
    HospitalLosGroupService.prototype.losLaneMax;
    /**
     * things that need to emit before generating the Hospital Los lanes
     * @type {?}
     */
    HospitalLosGroupService.prototype.dependency;
}
