// ng-filters-library/esm2015/lib/group/provider-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ProviderFilterService } from '../filter/providers/provider-filter.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../filter/providers/provider-filter.service";
export class ProviderGroupService {
    /**
     * @param {?} providerService
     */
    constructor(providerService) {
        this.providerService = providerService;
        /**
         * things that need to emit before generating the provider lanes
         */
        this.dependency = new Subject();
        this.providerService.providersLoaded.subscribe(() => {
            this.dependency.next();
        });
    }
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    generateLanes() {
        /** @type {?} */
        const option = {
            groupName: 'Provider',
            lanes: []
        };
        for (const provider of this.providerService.providers) {
            option.lanes.push({
                laneFunction: this.generateProviderLaneFilterFunction(provider),
                laneTitle: provider.lastName + ', ' + provider.firstName
            });
        }
        return option;
    }
    /**
     * helper function for generating a filter function for provider groups given a provider
     * @param {?} provider - provider to filter for
     * @return {?}
     */
    generateProviderLaneFilterFunction(provider) {
        /** @type {?} */
        const firstName = provider.firstName;
        /** @type {?} */
        const lastName = provider.lastName;
        return function (peraGraph) {
            return peraGraph.attendingProvider.firstName === firstName && peraGraph.attendingProvider.lastName === lastName;
        };
    }
}
ProviderGroupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ProviderGroupService.ctorParameters = () => [
    { type: ProviderFilterService }
];
/** @nocollapse */ ProviderGroupService.ngInjectableDef = i0.defineInjectable({ factory: function ProviderGroupService_Factory() { return new ProviderGroupService(i0.inject(i1.ProviderFilterService)); }, token: ProviderGroupService, providedIn: "root" });
if (false) {
    /**
     * things that need to emit before generating the provider lanes
     * @type {?}
     */
    ProviderGroupService.prototype.dependency;
    /** @type {?} */
    ProviderGroupService.prototype.providerService;
}
