// ng-filters-library/esm2015/lib/group/warning-category-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class WarningCategoryGroupService {
    constructor() {
        /**
         * things that need to emit before generating the warning lanes
         */
        this.dependency = null;
    }
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    generateLanes() {
        return {
            groupName: 'Warning Category',
            lanes: [{
                    laneFunction: (peraGraph) => peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'V',
                    laneTitle: 'Very High Warning'
                },
                {
                    laneFunction: (peraGraph) => peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'H',
                    laneTitle: 'High Warning'
                },
                {
                    laneFunction: (peraGraph) => peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'M',
                    laneTitle: 'Medium Warning'
                },
                { laneFunction: (peraGraph) => peraGraph.activeWarning == null, laneTitle: 'No Warning' }]
        };
    }
}
WarningCategoryGroupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
WarningCategoryGroupService.ctorParameters = () => [];
/** @nocollapse */ WarningCategoryGroupService.ngInjectableDef = i0.defineInjectable({ factory: function WarningCategoryGroupService_Factory() { return new WarningCategoryGroupService(); }, token: WarningCategoryGroupService, providedIn: "root" });
if (false) {
    /**
     * things that need to emit before generating the warning lanes
     * @type {?}
     */
    WarningCategoryGroupService.prototype.dependency;
}
