// ng-common-library/esm2015/lib/service/token/api/token-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { User } from '../../../model/User';
import { TokenConfigService } from '../config/token-config.service';
import { Storage } from '@ionic/storage';
import { first } from 'rxjs/operators';
import { JwtHelperService } from '@auth0/angular-jwt';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "@auth0/angular-jwt/src/jwthelper.service";
import * as i3 from "../config/token-config.service";
import * as i4 from "@ionic/storage";
export class TokenApiService {
    /**
     * @param {?} http
     * @param {?} jwtHelper
     * @param {?} config
     * @param {?} storage
     */
    constructor(http, jwtHelper, config, storage) {
        this.http = http;
        this.jwtHelper = jwtHelper;
        this.config = config;
        this.storage = storage;
        this.jwtTokenName = 'jwt_token';
        this.appId = null;
        this.deviceToken = null;
        this.token$ = new Subject();
    }
    /**
     * Check the current token.
     * @return {?}
     */
    checkToken() {
        this.token$.pipe(first()).subscribe((token) => {
            /** @type {?} */
            const decodedToken = this.jwtHelper.decodeToken(token);
            if (decodedToken) {
                /** @type {?} */
                const user = new User();
                user.username = decodedToken['sub'];
                user.roles = decodedToken['roles'];
                user.firstname = decodedToken['firstname'];
                user.lastname = decodedToken['lastname'];
                user.appid = decodedToken['appid'];
                // console.log('Existing, valid client side token found. Checking the server.');
                this.http.post(this.config.getCheckTokenEndPoint(), user).pipe(first()).subscribe(() => {
                    // If we land in here, then the token interceptor on the server validated the token
                    this.appId = user.appid;
                    // console.log('Valid token!');
                    this.token$.next(token);
                }, (e) => {
                    this.appId = null;
                    // console.log('Invalid token!');
                    this.token$.next(null);
                });
            }
            else {
                // console.log('Removing invalid token from storage.');
                this.clearToken();
                this.token$.next(null);
            }
        });
        this.getExistingToken();
    }
    /**
     * Refresh the token for the given user.
     *
     * @param {?} user User for which to refresh token.
     * @return {?}
     */
    refreshToken(user) {
        if (!user || !user.username) {
            // console.log('Valid user not supplied, cannot refresh token!');
        }
        else {
            setTimeout(() => {
                this.token$.pipe(first()).subscribe((token) => {
                    // Check whether or not the token is valid according to the server
                    this.http.post(this.config.getRefreshTokenEndPoint(), user, { responseType: 'text' }).pipe(first()).subscribe((token) => {
                        // console.log('Successfully retrieved updated token [' + token + ']');
                        this.storage.set(this.jwtTokenName, token);
                    }, () => {
                        // The server thinks the token is not valid (expired, the secret key changed, etc)
                        // console.log('Invalid token!');
                        this.clearToken();
                        this.token$.next(null);
                    });
                });
                this.getExistingToken();
            }, 100);
        }
    }
    /**
     * @param {?} user
     * @return {?}
     */
    getToken(user) {
        // Check whether or not the token is valid according to the server
        this.http.post(this.config.getGetTokenEndPoint(), user, { responseType: 'text' }).pipe(first()).subscribe((token) => {
            // console.log('Successfully retrieved new token: [' + token + ']');
            this.appId = user.appid;
            this.storage.set(this.jwtTokenName, token)
                .then(() => {
                // console.log('New token added to storage. [' + token + ']')
                this.token$.next(token);
            }, () => {
                // console.log('Could not store new token!');
                this.token$.next(null);
            });
        }, (e) => {
            // The server thinks the token is not valid (expired, the secret key changed, etc)
            // console.log('Could not retrieve new token!');
            this.token$.next(null);
        });
    }
    /**
     * @return {?}
     */
    clearToken() {
        this.storage.remove(this.jwtTokenName)
            .then((jwt) => {
            // console.log('JWT removed from storage');
        }, (reason) => {
            // console.log('JWT could not be removed from storage. Reason = ' + reason);
        });
    }
    /**
     * @return {?}
     */
    getExistingToken() {
        this.storage.get(this.jwtTokenName)
            .then((token) => {
            this.token$.next(token);
        }, (error) => {
            this.token$.next(null);
        })
            .catch((error) => {
            this.token$.next(null);
        });
    }
    /**
     * @return {?}
     */
    getAppId() {
        return this.appId;
    }
}
TokenApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TokenApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: JwtHelperService },
    { type: TokenConfigService },
    { type: Storage }
];
/** @nocollapse */ TokenApiService.ngInjectableDef = i0.defineInjectable({ factory: function TokenApiService_Factory() { return new TokenApiService(i0.inject(i1.HttpClient), i0.inject(i2.JwtHelperService), i0.inject(i3.TokenConfigService), i0.inject(i4.Storage)); }, token: TokenApiService, providedIn: "root" });
if (false) {
    /** @type {?} */
    TokenApiService.prototype.jwtTokenName;
    /** @type {?} */
    TokenApiService.prototype.appId;
    /** @type {?} */
    TokenApiService.prototype.deviceToken;
    /** @type {?} */
    TokenApiService.prototype.token$;
    /** @type {?} */
    TokenApiService.prototype.http;
    /** @type {?} */
    TokenApiService.prototype.jwtHelper;
    /** @type {?} */
    TokenApiService.prototype.config;
    /** @type {?} */
    TokenApiService.prototype.storage;
}
