// ng-common-library/esm2015/lib/service/diagnosis/config/diagnosis-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DiagnosisConfigService {
    constructor() {
        this.primaryDiagnosesEndpoint = null;
    }
    /**
     * @return {?}
     */
    getPrimaryDiagnosesEndpoint() {
        if (!this.primaryDiagnosesEndpoint) {
            console.warn('Primary diagnoses endpoint is not set!');
        }
        return this.primaryDiagnosesEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setPrimaryDiagnosesEndpoint(value) {
        this.primaryDiagnosesEndpoint = value;
    }
}
DiagnosisConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ DiagnosisConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisConfigService_Factory() { return new DiagnosisConfigService(); }, token: DiagnosisConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DiagnosisConfigService.prototype.primaryDiagnosesEndpoint;
}
