// ng-filters-library/esm5/lib/filter/agegroups/age-groups.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FilterService } from '../filter.service';
import * as i0 from "@angular/core";
import * as i1 from "../filter.service";
var AgeGroupsService = /** @class */ (function () {
    function AgeGroupsService(filterService) {
        this.filterService = filterService;
        this.PEDIATRIC_GROUP = 'Pediatric';
        this.ADULT_GROUP = 'Adult';
        this.ALL_GROUP = 'All';
    }
    /**
     * @return {?}
     */
    AgeGroupsService.prototype.unstageAgeGroups = /**
     * @return {?}
     */
    function () {
        this.filterService.unstageFilter(this.PEDIATRIC_GROUP);
        this.filterService.unstageFilter(this.ADULT_GROUP);
        this.filterService.unstageFilter(this.ALL_GROUP);
    };
    /**
     * Always returns true because we want all the patients
     * @param {?} peragraph - peragraph to filter
     * @return {?}
     */
    AgeGroupsService.prototype.allFilterFunction = /**
     * Always returns true because we want all the patients
     * @param {?} peragraph - peragraph to filter
     * @return {?}
     */
    function (peragraph) {
        return true;
    };
    /**
     * Filter out any peragraphs with patients younger than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    AgeGroupsService.prototype.adultFilterFunction = /**
     * Filter out any peragraphs with patients younger than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    function (peragraph) {
        /** @type {?} */
        var birthDate = new Date();
        birthDate.setMilliseconds(peragraph.patient.dob);
        /** @type {?} */
        var currentDate = new Date();
        currentDate.setMilliseconds(Date.now());
        /** @type {?} */
        var age = currentDate.getFullYear() - birthDate.getFullYear();
        return age >= 18;
    };
    /**
     * Filter out any peragraphs with patients older than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    AgeGroupsService.prototype.pediatricFilterFunction = /**
     * Filter out any peragraphs with patients older than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    function (peragraph) {
        /** @type {?} */
        var birthDate = new Date();
        birthDate.setMilliseconds(peragraph.patient.dob);
        /** @type {?} */
        var currentDate = new Date();
        currentDate.setMilliseconds(Date.now());
        /** @type {?} */
        var age = currentDate.getFullYear() - birthDate.getFullYear();
        return age < 18;
    };
    AgeGroupsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AgeGroupsService.ctorParameters = function () { return [
        { type: FilterService }
    ]; };
    /** @nocollapse */ AgeGroupsService.ngInjectableDef = i0.defineInjectable({ factory: function AgeGroupsService_Factory() { return new AgeGroupsService(i0.inject(i1.FilterService)); }, token: AgeGroupsService, providedIn: "root" });
    return AgeGroupsService;
}());
export { AgeGroupsService };
if (false) {
    /** @type {?} */
    AgeGroupsService.prototype.PEDIATRIC_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.ADULT_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.ALL_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.filterService;
}
