// ng-common-library/esm2015/lib/service/peragraph/config/peragraph-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PeragraphConfigService {
    constructor() {
        this.singlePeraGraphEndPoint = null;
        this.allPeraGraphsEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setSinglePeraGraphEndPoint(value) {
        this.singlePeraGraphEndPoint = value;
    }
    /**
     * @return {?}
     */
    getSinglePeraGraphEndPoint() {
        if (!this.singlePeraGraphEndPoint) {
            console.warn('Single PeraGraph endpoint is not set!');
        }
        return this.singlePeraGraphEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setAllPeraGraphsEndPoint(value) {
        this.allPeraGraphsEndPoint = value;
    }
    /**
     * @return {?}
     */
    getAllPeraGraphsEndPoint() {
        if (!this.allPeraGraphsEndPoint) {
            console.warn('All PeraGraph endpoint is not set!');
        }
        return this.allPeraGraphsEndPoint;
    }
}
PeragraphConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ PeragraphConfigService.ngInjectableDef = i0.defineInjectable({ factory: function PeragraphConfigService_Factory() { return new PeragraphConfigService(); }, token: PeragraphConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    PeragraphConfigService.prototype.singlePeraGraphEndPoint;
    /** @type {?} */
    PeragraphConfigService.prototype.allPeraGraphsEndPoint;
}
