// ng-common-library/esm2015/lib/service/locations/config/locations-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class LocationsConfigService {
    constructor() {
        this.activeFacilitiesEndpoint = null;
        this.activeUnitsEndpoint = null;
    }
    /**
     * @return {?}
     */
    getActiveFacilitiesEndpoint() {
        if (!this.activeFacilitiesEndpoint) {
            console.warn('Active facilities endpoint is not set!');
        }
        return this.activeFacilitiesEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setActiveFacilitiesEndpoint(value) {
        this.activeFacilitiesEndpoint = value;
    }
    /**
     * @return {?}
     */
    getActiveUnitsEndpoint() {
        if (!this.activeUnitsEndpoint) {
            console.warn('Active units endpoint is not set!');
        }
        return this.activeUnitsEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setActiveUnitsEndpoint(value) {
        this.activeUnitsEndpoint = value;
    }
}
LocationsConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ LocationsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function LocationsConfigService_Factory() { return new LocationsConfigService(); }, token: LocationsConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    LocationsConfigService.prototype.activeFacilitiesEndpoint;
    /** @type {?} */
    LocationsConfigService.prototype.activeUnitsEndpoint;
}
