// ng-filters-library/esm5/lib/filter/providers/providers-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ProvidersConfigService = /** @class */ (function () {
    function ProvidersConfigService() {
        this.allProvidersEndpoint = null;
    }
    /**
     * @return {?}
     */
    ProvidersConfigService.prototype.getAllProvidersEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.allProvidersEndpoint) {
            console.warn('All Providers Endpoint is not set!');
        }
        return this.allProvidersEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ProvidersConfigService.prototype.setAllProvidersEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.allProvidersEndpoint = value;
    };
    ProvidersConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ProvidersConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ProvidersConfigService_Factory() { return new ProvidersConfigService(); }, token: ProvidersConfigService, providedIn: "root" });
    return ProvidersConfigService;
}());
export { ProvidersConfigService };
if (false) {
    /** @type {?} */
    ProvidersConfigService.prototype.allProvidersEndpoint;
}
