// ng-filters-library/esm5/lib/filter/providers/providers-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from "@angular/common/http";
import { ProvidersConfigService } from './providers-config.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./providers-config.service";
var ProvidersApiService = /** @class */ (function () {
    function ProvidersApiService(http, providersConfigService) {
        this.http = http;
        this.providersConfigService = providersConfigService;
    }
    /**
     * @return {?}
     */
    ProvidersApiService.prototype.getAllProviders = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.providersConfigService.getAllProvidersEndpoint());
    };
    ProvidersApiService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ProvidersApiService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: ProvidersConfigService }
    ]; };
    /** @nocollapse */ ProvidersApiService.ngInjectableDef = i0.defineInjectable({ factory: function ProvidersApiService_Factory() { return new ProvidersApiService(i0.inject(i1.HttpClient), i0.inject(i2.ProvidersConfigService)); }, token: ProvidersApiService, providedIn: "root" });
    return ProvidersApiService;
}());
export { ProvidersApiService };
if (false) {
    /** @type {?} */
    ProvidersApiService.prototype.http;
    /** @type {?} */
    ProvidersApiService.prototype.providersConfigService;
}
