// ng-filters-library/esm5/lib/filter/providers/provider-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ProviderFilterService = /** @class */ (function () {
    function ProviderFilterService() {
        // all providers
        this.providers = [];
        // dictionary of providers currently selected in sidebar
        this.selectedProviders = {};
        // event for letting provider filter know providers are ready
        this.providersLoaded = new EventEmitter();
        // event for changing all checkbox values
        // true for select
        // false for deselect
        this.toggleAll = new EventEmitter();
    }
    /**
     * closure which returns filter function
     * had to use a closure to be able to pass in selectedProviders list into filter
     */
    /**
     * closure which returns filter function
     * had to use a closure to be able to pass in selectedProviders list into filter
     * @return {?}
     */
    ProviderFilterService.prototype.getFilterfunction = /**
     * closure which returns filter function
     * had to use a closure to be able to pass in selectedProviders list into filter
     * @return {?}
     */
    function () {
        /** @type {?} */
        var selectedProviders = {};
        for (var prov in this.selectedProviders) {
            if (prov) {
                selectedProviders[prov] = this.selectedProviders[prov];
            }
        }
        return function (peragraph) {
            return peragraph.attendingProvider && selectedProviders[peragraph.attendingProvider.firstName + peragraph.attendingProvider.lastName];
        };
    };
    ProviderFilterService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ProviderFilterService.ctorParameters = function () { return []; };
    /** @nocollapse */ ProviderFilterService.ngInjectableDef = i0.defineInjectable({ factory: function ProviderFilterService_Factory() { return new ProviderFilterService(); }, token: ProviderFilterService, providedIn: "root" });
    return ProviderFilterService;
}());
export { ProviderFilterService };
if (false) {
    /** @type {?} */
    ProviderFilterService.prototype.providers;
    /** @type {?} */
    ProviderFilterService.prototype.selectedProviders;
    /** @type {?} */
    ProviderFilterService.prototype.providersLoaded;
    /** @type {?} */
    ProviderFilterService.prototype.toggleAll;
}
