// ng-common-library/esm2015/lib/service/flowsheets/config/flowsheets-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class FlowsheetsConfigService {
    constructor() {
        this.singlePeraGraphFlowSheetEndPoint = null;
        this.allPeragraphFlowSheetsEndPoint = null;
        this.detailedFlowSheetColumnEndpoint = null;
        this.latestRiScoresEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setSinglePeraGraphFlowSheetEndPoint(value) {
        this.singlePeraGraphFlowSheetEndPoint = value;
    }
    /**
     * @return {?}
     */
    getSinglePeraGraphFlowSheetEndPoint() {
        if (!this.singlePeraGraphFlowSheetEndPoint) {
            console.warn('Single PeraGraph flowsheet endpoint is not set!');
        }
        return this.singlePeraGraphFlowSheetEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setAllPeragraphFlowSheetsEndPoint(value) {
        this.allPeragraphFlowSheetsEndPoint = value;
    }
    /**
     * @return {?}
     */
    getAllPeragraphFlowSheetsEndPoint() {
        if (!this.allPeragraphFlowSheetsEndPoint) {
            console.warn('All PeraGraph flowsheets endpoint is not set!');
        }
        return this.allPeragraphFlowSheetsEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setDetailedFlowSheetColumnEndpoint(value) {
        this.detailedFlowSheetColumnEndpoint = value;
    }
    /**
     * @return {?}
     */
    getDetailedFlowSheetColumnEndpoint() {
        if (!this.detailedFlowSheetColumnEndpoint) {
            console.warn('Detailed flowsheet column endpoint is not set!');
        }
        return this.detailedFlowSheetColumnEndpoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setLatestRiScoresEndPoint(value) {
        this.latestRiScoresEndPoint = value;
    }
    /**
     * @return {?}
     */
    getLatestRiScoresEndPoint() {
        if (!this.latestRiScoresEndPoint) {
            console.warn('Latest RI Scores endpoint is not set!');
        }
        return this.latestRiScoresEndPoint;
    }
}
FlowsheetsConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ FlowsheetsConfigService.ngInjectableDef = i0.defineInjectable({ factory: function FlowsheetsConfigService_Factory() { return new FlowsheetsConfigService(); }, token: FlowsheetsConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    FlowsheetsConfigService.prototype.singlePeraGraphFlowSheetEndPoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.allPeragraphFlowSheetsEndPoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.detailedFlowSheetColumnEndpoint;
    /** @type {?} */
    FlowsheetsConfigService.prototype.latestRiScoresEndPoint;
}
