// ng-common-library/esm2015/lib/service/flowsheets/api/flowsheets-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { FlowsheetsConfigService } from '../config/flowsheets-config.service';
import { BehaviorSubject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/flowsheets-config.service";
export class FlowsheetsApiService {
    /**
     * @param {?} http
     * @param {?} flowsheetConfigService
     */
    constructor(http, flowsheetConfigService) {
        this.http = http;
        this.flowsheetConfigService = flowsheetConfigService;
        this.pollInstance$ = timer(0, 60000);
        this.flowSheetSubject = new BehaviorSubject(null);
        this.riScoresSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    getFlowSheetSubject() {
        return this.flowSheetSubject;
    }
    /**
     * @return {?}
     */
    getFlowSheetSubjectValue() {
        return this.flowSheetSubject.value;
    }
    /**
     * @return {?}
     */
    getRiScoresSubject() {
        return this.riScoresSubject;
    }
    /**
     * @return {?}
     */
    getRiScoresSubjectValue() {
        return this.riScoresSubject.value;
    }
    /**
     * @return {?}
     */
    getSinglePeraGraphFlowSheet() {
        return this.http.get(this.flowsheetConfigService.getSinglePeraGraphFlowSheetEndPoint());
    }
    /**
     * @return {?}
     */
    getDetailedFlowsheetColumn() {
        return this.http.get(this.flowsheetConfigService.getDetailedFlowSheetColumnEndpoint());
    }
    /**
     * @return {?}
     */
    getAllPeraGraphFlowSheets() {
        return this.http.get(this.flowsheetConfigService.getAllPeragraphFlowSheetsEndPoint(), { observe: 'response' });
    }
    /**
     * @return {?}
     */
    getAllRiScores() {
        return this.http.get(this.flowsheetConfigService.getLatestRiScoresEndPoint(), { observe: 'response' });
    }
    /**
     * @return {?}
     */
    getFlowSheetInterval() {
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(() => this.getAllPeraGraphFlowSheets()));
    }
    /**
     * @return {?}
     */
    getLatestRiScoresInterval() {
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(() => this.getAllRiScores()));
    }
}
FlowsheetsApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
FlowsheetsApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: FlowsheetsConfigService }
];
/** @nocollapse */ FlowsheetsApiService.ngInjectableDef = i0.defineInjectable({ factory: function FlowsheetsApiService_Factory() { return new FlowsheetsApiService(i0.inject(i1.HttpClient), i0.inject(i2.FlowsheetsConfigService)); }, token: FlowsheetsApiService, providedIn: "root" });
if (false) {
    /** @type {?} */
    FlowsheetsApiService.prototype.flowSheetSubject;
    /** @type {?} */
    FlowsheetsApiService.prototype.riScoresSubject;
    /** @type {?} */
    FlowsheetsApiService.prototype.pollInstance$;
    /** @type {?} */
    FlowsheetsApiService.prototype.http;
    /** @type {?} */
    FlowsheetsApiService.prototype.flowsheetConfigService;
}
