// ng-common-library/esm2015/lib/service/chart.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MarkerColors } from '../model/MarkerColors';
import { DateService } from './date.service';
import * as i0 from "@angular/core";
import * as i1 from "./date.service";
export class ChartService {
    /**
     * @param {?} dateService
     */
    constructor(dateService) {
        this.dateService = dateService;
    }
    /**
     * Create the line across the 0 axis for when RI scores go below zero in a graph.
     * @param {?} minRiScore Minimum RI score.
     * @return {?}
     */
    createZeroAxisLine(minRiScore) {
        return [
            {
                color: '#000000',
                width: minRiScore < 0 ? 1 : null,
                value: minRiScore < 0 ? 0 : null,
                zIndex: 3
            }
        ];
    }
    /**
     * Create an alert plot line for a graph.
     *
     * @param {?} graph
     * @return {?}
     */
    updateAlertPlotLine(graph) {
        /** @type {?} */
        const plotLines = [];
        if (graph && graph.activeWarning) {
            plotLines.push({
                color: '#CE2029',
                value: this.dateService.translateDate(graph.activeWarning.createdDate, graph.timeZone),
                width: 2,
                zIndex: 3
            });
        }
        return plotLines;
    }
    /**
     * @param {?} minRiScore
     * @return {?}
     */
    getMinYAxisTick(minRiScore) {
        return (minRiScore > 0) ? 0 : 20 * Math.floor(minRiScore / 20);
    }
    /**
     * Used in the graph to depict the unit type where a health score was recorded.
     * @param {?} unitType Type of unit to get the marker color for.
     * @return {?}
     */
    getMarkerColor(unitType) {
        /** @type {?} */
        let markerColor = MarkerColors.Normal;
        switch (unitType) {
            case 'N':
                markerColor = MarkerColors.IntensiveCare;
                break;
            case 'C':
                markerColor = MarkerColors.IntensiveCare;
                break;
            case 'P':
                markerColor = MarkerColors.StepDown;
                break;
        }
        return markerColor;
    }
}
ChartService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ChartService.ctorParameters = () => [
    { type: DateService }
];
/** @nocollapse */ ChartService.ngInjectableDef = i0.defineInjectable({ factory: function ChartService_Factory() { return new ChartService(i0.inject(i1.DateService)); }, token: ChartService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ChartService.prototype.dateService;
}
