// ng-array-view-library/esm5/lib/group/pera-graph/service/pera-graph.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WarningColors } from '../../../model/WarningColors';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var PeraGraphService = /** @class */ (function () {
    function PeraGraphService() {
        // Subject for navigating to the PSG
        this.graphClicked$ = new Subject();
    }
    /**
     * @param {?} graph
     * @return {?}
     */
    PeraGraphService.prototype.graphClickedEvent = /**
     * @param {?} graph
     * @return {?}
     */
    function (graph) {
        this.closeToolTip();
        this.graphClicked$.next(graph);
    };
    /**
     * @param {?} warning
     * @param {?} latestRi
     * @param {?} isDischarged
     * @return {?}
     */
    PeraGraphService.prototype.getBackgroundColorByWarningAndRi = /**
     * @param {?} warning
     * @param {?} latestRi
     * @param {?} isDischarged
     * @return {?}
     */
    function (warning, latestRi, isDischarged) {
        /** @type {?} */
        var backgroundColor;
        if (warning && !isDischarged) {
            switch (warning.category) {
                case 'V':
                    backgroundColor = WarningColors.VeryHigh;
                    break;
                case 'H':
                    backgroundColor = WarningColors.High;
                    break;
                case 'M':
                    backgroundColor = WarningColors.Medium;
                    break;
            }
        }
        else if (latestRi && !isDischarged) {
            backgroundColor = WarningColors.NoCurrent;
        }
        else {
            backgroundColor = WarningColors.NoRi;
        }
        return backgroundColor;
    };
    /**
     * @param {?} latestRi
     * @return {?}
     */
    PeraGraphService.prototype.getRiScoreToDisplay = /**
     * @param {?} latestRi
     * @return {?}
     */
    function (latestRi) {
        return (latestRi) ? latestRi + '' : '--';
    };
    /**
     * @return {?}
     */
    PeraGraphService.prototype.closeToolTip = /**
     * @return {?}
     */
    function () {
        if (this.tooltip && this.tooltip.isOpen) {
            this.tooltip.close();
        }
    };
    PeraGraphService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PeraGraphService.ctorParameters = function () { return []; };
    /** @nocollapse */ PeraGraphService.ngInjectableDef = i0.defineInjectable({ factory: function PeraGraphService_Factory() { return new PeraGraphService(); }, token: PeraGraphService, providedIn: "root" });
    return PeraGraphService;
}());
export { PeraGraphService };
if (false) {
    /** @type {?} */
    PeraGraphService.prototype.graphClicked$;
    /** @type {?} */
    PeraGraphService.prototype.tooltip;
}
