// ng-array-view-library/esm5/lib/group/pera-graph/service/tooltip.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ArrayViewService } from '../../../service/array-view.service';
import { first } from 'rxjs/operators';
import { PeraGraphService } from './pera-graph.service';
import { ArrayViewConfigService } from '../../../service/array-view-config.service';
import { LoginService } from 'ng-login-library';
import { TokenApiService } from 'ng-common-library';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./pera-graph.service";
import * as i2 from "../../../service/array-view.service";
import * as i3 from "@angular/common/http";
import * as i4 from "ng-login-library";
import * as i5 from "ng-common-library";
import * as i6 from "../../../service/array-view-config.service";
var TooltipService = /** @class */ (function () {
    function TooltipService(graphService, arrayViewService, http, loginService, tokenApiService, arrayViewConfigService) {
        this.graphService = graphService;
        this.arrayViewService = arrayViewService;
        this.http = http;
        this.loginService = loginService;
        this.tokenApiService = tokenApiService;
        this.arrayViewConfigService = arrayViewConfigService;
        this.toolTipInfoUpdated$ = new Subject();
    }
    /**
     * @param {?} graph
     * @return {?}
     */
    TooltipService.prototype.updateToolTipInfo = /**
     * @param {?} graph
     * @return {?}
     */
    function (graph) {
        var _this = this;
        this.tokenApiService.refreshToken(this.loginService.getCurrentUser());
        // Get the current tooltip info from the server.
        this.http.get(this.arrayViewConfigService.getToolTipInfoEndPoint() + '/pid/' + graph.patient.pid + '/visit/' + graph.visitId).pipe(first()).subscribe(function (toolTipInfo) {
            // Ensure that the user hasn't moved onto a new graph with a different during the time that the call was made to the server
            if (graph.visitId === toolTipInfo.visitId) {
                /** @type {?} */
                var numberOfOtherVisitsWithValidRi = toolTipInfo.numberOfOtherVisitsWithValidRi;
                // The query will return a -1 value if the current and only visit has no valid RI scores.
                graph.numberVisitsWithValidRI = (numberOfOtherVisitsWithValidRi > 0) ? numberOfOtherVisitsWithValidRi : 0;
                // Tooltip diagnosis formatting
                if (toolTipInfo.diagnosis != null) {
                    // Replace the first ^ with nothing
                    /** @type {?} */
                    var tmpDiag = toolTipInfo.diagnosis.replace('^', '');
                    // Replace the rest of the ^ with ', '
                    graph.diagnosis = tmpDiag.replace(/\^/g, ', ');
                }
                _this.toolTipInfoUpdated$.next(toolTipInfo);
            }
        });
    };
    TooltipService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TooltipService.ctorParameters = function () { return [
        { type: PeraGraphService },
        { type: ArrayViewService },
        { type: HttpClient },
        { type: LoginService },
        { type: TokenApiService },
        { type: ArrayViewConfigService }
    ]; };
    /** @nocollapse */ TooltipService.ngInjectableDef = i0.defineInjectable({ factory: function TooltipService_Factory() { return new TooltipService(i0.inject(i1.PeraGraphService), i0.inject(i2.ArrayViewService), i0.inject(i3.HttpClient), i0.inject(i4.LoginService), i0.inject(i5.TokenApiService), i0.inject(i6.ArrayViewConfigService)); }, token: TooltipService, providedIn: "root" });
    return TooltipService;
}());
export { TooltipService };
if (false) {
    /** @type {?} */
    TooltipService.prototype.toolTipInfoUpdated$;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.graphService;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.arrayViewService;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.loginService;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.tokenApiService;
    /**
     * @type {?}
     * @private
     */
    TooltipService.prototype.arrayViewConfigService;
}
