// ng-common-library/esm5/lib/service/peragraph/config/peragraph-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var PeragraphConfigService = /** @class */ (function () {
    function PeragraphConfigService() {
        this.singlePeraGraphEndPoint = null;
        this.allPeraGraphsEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    PeragraphConfigService.prototype.setSinglePeraGraphEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.singlePeraGraphEndPoint = value;
    };
    /**
     * @return {?}
     */
    PeragraphConfigService.prototype.getSinglePeraGraphEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.singlePeraGraphEndPoint) {
            console.warn('Single PeraGraph endpoint is not set!');
        }
        return this.singlePeraGraphEndPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    PeragraphConfigService.prototype.setAllPeraGraphsEndPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.allPeraGraphsEndPoint = value;
    };
    /**
     * @return {?}
     */
    PeragraphConfigService.prototype.getAllPeraGraphsEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.allPeraGraphsEndPoint) {
            console.warn('All PeraGraph endpoint is not set!');
        }
        return this.allPeraGraphsEndPoint;
    };
    PeragraphConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ PeragraphConfigService.ngInjectableDef = i0.defineInjectable({ factory: function PeragraphConfigService_Factory() { return new PeragraphConfigService(); }, token: PeragraphConfigService, providedIn: "root" });
    return PeragraphConfigService;
}());
export { PeragraphConfigService };
if (false) {
    /** @type {?} */
    PeragraphConfigService.prototype.singlePeraGraphEndPoint;
    /** @type {?} */
    PeragraphConfigService.prototype.allPeraGraphsEndPoint;
}
