// ng-common-library/esm5/lib/service/usage/config/usage-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var UsageConfigService = /** @class */ (function () {
    function UsageConfigService() {
        this.usagePollingInterval = 5000;
    }
    /**
     * @param {?} value Endpoint for sending usage data.
     * @return {?}
     */
    UsageConfigService.prototype.setUsageDataEndPoint = /**
     * @param {?} value Endpoint for sending usage data.
     * @return {?}
     */
    function (value) {
        this.usageDataEndPoint = value;
    };
    /**
     * Endpoint for sending usage data.
     * @return {?}
     */
    UsageConfigService.prototype.getUsageDataEndPoint = /**
     * Endpoint for sending usage data.
     * @return {?}
     */
    function () {
        if (!this.usageDataEndPoint) {
            console.warn('usage data endpoint is not set!');
        }
        return this.usageDataEndPoint;
    };
    /**
     * Number of milliseconds between polling attempts.
     * @return {?}
     */
    UsageConfigService.prototype.getUsagePollingInterval = /**
     * Number of milliseconds between polling attempts.
     * @return {?}
     */
    function () {
        return this.usagePollingInterval;
    };
    /**
     * @param {?} value Number of milliseconds between polling attempts.
     * @return {?}
     */
    UsageConfigService.prototype.setUsagePollingInterval = /**
     * @param {?} value Number of milliseconds between polling attempts.
     * @return {?}
     */
    function (value) {
        this.usagePollingInterval = value;
    };
    UsageConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ UsageConfigService.ngInjectableDef = i0.defineInjectable({ factory: function UsageConfigService_Factory() { return new UsageConfigService(); }, token: UsageConfigService, providedIn: "root" });
    return UsageConfigService;
}());
export { UsageConfigService };
if (false) {
    /** @type {?} */
    UsageConfigService.prototype.usageDataEndPoint;
    /** @type {?} */
    UsageConfigService.prototype.usagePollingInterval;
}
