// ng-filters-library/esm2015/lib/filter/agegroups/age-groups.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FilterService } from '../filter.service';
import * as i0 from "@angular/core";
import * as i1 from "../filter.service";
export class AgeGroupsService {
    /**
     * @param {?} filterService
     */
    constructor(filterService) {
        this.filterService = filterService;
        this.PEDIATRIC_GROUP = 'Pediatric';
        this.ADULT_GROUP = 'Adult';
        this.ALL_GROUP = 'All';
    }
    /**
     * @return {?}
     */
    unstageAgeGroups() {
        this.filterService.unstageFilter(this.PEDIATRIC_GROUP);
        this.filterService.unstageFilter(this.ADULT_GROUP);
        this.filterService.unstageFilter(this.ALL_GROUP);
    }
    /**
     * Always returns true because we want all the patients
     * @param {?} peragraph - peragraph to filter
     * @return {?}
     */
    allFilterFunction(peragraph) {
        return true;
    }
    /**
     * Filter out any peragraphs with patients younger than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    adultFilterFunction(peragraph) {
        /** @type {?} */
        const birthDate = new Date();
        birthDate.setMilliseconds(peragraph.patient.dob);
        /** @type {?} */
        const currentDate = new Date();
        currentDate.setMilliseconds(Date.now());
        /** @type {?} */
        const age = currentDate.getFullYear() - birthDate.getFullYear();
        return age >= 18;
    }
    /**
     * Filter out any peragraphs with patients older than 18
     * @param {?} peragraph - peragraph to be filtered
     * @return {?}
     */
    pediatricFilterFunction(peragraph) {
        /** @type {?} */
        const birthDate = new Date();
        birthDate.setMilliseconds(peragraph.patient.dob);
        /** @type {?} */
        const currentDate = new Date();
        currentDate.setMilliseconds(Date.now());
        /** @type {?} */
        const age = currentDate.getFullYear() - birthDate.getFullYear();
        return age < 18;
    }
}
AgeGroupsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AgeGroupsService.ctorParameters = () => [
    { type: FilterService }
];
/** @nocollapse */ AgeGroupsService.ngInjectableDef = i0.defineInjectable({ factory: function AgeGroupsService_Factory() { return new AgeGroupsService(i0.inject(i1.FilterService)); }, token: AgeGroupsService, providedIn: "root" });
if (false) {
    /** @type {?} */
    AgeGroupsService.prototype.PEDIATRIC_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.ADULT_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.ALL_GROUP;
    /** @type {?} */
    AgeGroupsService.prototype.filterService;
}
