// ng-common-library/esm2015/lib/service/usage/config/usage-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class UsageConfigService {
    constructor() {
        this.usagePollingInterval = 5000;
    }
    /**
     * @param {?} value Endpoint for sending usage data.
     * @return {?}
     */
    setUsageDataEndPoint(value) {
        this.usageDataEndPoint = value;
    }
    /**
     * Endpoint for sending usage data.
     * @return {?}
     */
    getUsageDataEndPoint() {
        if (!this.usageDataEndPoint) {
            console.warn('usage data endpoint is not set!');
        }
        return this.usageDataEndPoint;
    }
    /**
     * Number of milliseconds between polling attempts.
     * @return {?}
     */
    getUsagePollingInterval() {
        return this.usagePollingInterval;
    }
    /**
     * @param {?} value Number of milliseconds between polling attempts.
     * @return {?}
     */
    setUsagePollingInterval(value) {
        this.usagePollingInterval = value;
    }
}
UsageConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ UsageConfigService.ngInjectableDef = i0.defineInjectable({ factory: function UsageConfigService_Factory() { return new UsageConfigService(); }, token: UsageConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    UsageConfigService.prototype.usageDataEndPoint;
    /** @type {?} */
    UsageConfigService.prototype.usagePollingInterval;
}
