// ng-filters-library/esm5/lib/filter/service-line/service-line-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ServiceLineConfigService = /** @class */ (function () {
    function ServiceLineConfigService() {
        this.serviceLineDescriptionsEndpoint = null;
    }
    /**
     * @return {?}
     */
    ServiceLineConfigService.prototype.getServiceLineDescriptionsEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.serviceLineDescriptionsEndpoint) {
            console.warn('Service Line Descriptions endpoint is not set!');
        }
        return this.serviceLineDescriptionsEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ServiceLineConfigService.prototype.setServiceLineDescriptionsEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.serviceLineDescriptionsEndpoint = value;
    };
    ServiceLineConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ServiceLineConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ServiceLineConfigService_Factory() { return new ServiceLineConfigService(); }, token: ServiceLineConfigService, providedIn: "root" });
    return ServiceLineConfigService;
}());
export { ServiceLineConfigService };
if (false) {
    /** @type {?} */
    ServiceLineConfigService.prototype.serviceLineDescriptionsEndpoint;
}
