// ng-filters-library/esm5/lib/filter/service-line/service-line-rest.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ServiceLineConfigService } from './service-line-config.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./service-line-config.service";
var ServiceLineRestService = /** @class */ (function () {
    function ServiceLineRestService(http, serviceLineConfig) {
        this.http = http;
        this.serviceLineConfig = serviceLineConfig;
    }
    /**
     * return all service line descriptions
     */
    /**
     * return all service line descriptions
     * @return {?}
     */
    ServiceLineRestService.prototype.getAllServiceLineDescriptions = /**
     * return all service line descriptions
     * @return {?}
     */
    function () {
        return this.http.get(this.serviceLineConfig.getServiceLineDescriptionsEndpoint());
    };
    ServiceLineRestService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ServiceLineRestService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: ServiceLineConfigService }
    ]; };
    /** @nocollapse */ ServiceLineRestService.ngInjectableDef = i0.defineInjectable({ factory: function ServiceLineRestService_Factory() { return new ServiceLineRestService(i0.inject(i1.HttpClient), i0.inject(i2.ServiceLineConfigService)); }, token: ServiceLineRestService, providedIn: "root" });
    return ServiceLineRestService;
}());
export { ServiceLineRestService };
if (false) {
    /** @type {?} */
    ServiceLineRestService.prototype.http;
    /** @type {?} */
    ServiceLineRestService.prototype.serviceLineConfig;
}
