// ng-filters-library/esm5/lib/group/facility-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { LocationSelectorService } from '../filter/locations/location-selector.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../filter/locations/location-selector.service";
var FacilityGroupService = /** @class */ (function () {
    function FacilityGroupService(locationSelectorService) {
        var _this = this;
        this.locationSelectorService = locationSelectorService;
        /**
         * things that need to emit before generating the facility lanes
         */
        this.dependency = new Subject();
        this.locationSelectorService.loadedLocations.subscribe(function () {
            _this.dependency.next();
        });
    }
    /**
     * generate group choice object for facilities
     */
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    FacilityGroupService.prototype.generateLanes = /**
     * generate group choice object for facilities
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {
            groupName: 'Facility',
            lanes: []
        };
        try {
            for (var _a = tslib_1.__values(this.locationSelectorService.facilities), _b = _a.next(); !_b.done; _b = _a.next()) {
                var facility = _b.value;
                option.lanes.push({ laneFunction: this.generateFacilitiyLaneFilterFunction(facility), laneTitle: facility.displayName });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return option;
        var e_1, _c;
    };
    /**
     * generate filter function for grouping by facility
     * @param {?} facility - facility to filter for
     * @return {?}
     */
    FacilityGroupService.prototype.generateFacilitiyLaneFilterFunction = /**
     * generate filter function for grouping by facility
     * @param {?} facility - facility to filter for
     * @return {?}
     */
    function (facility) {
        /** @type {?} */
        var facilityName = facility.displayName;
        return function (peraGraph) {
            return peraGraph.facilityId === facilityName;
        };
    };
    FacilityGroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FacilityGroupService.ctorParameters = function () { return [
        { type: LocationSelectorService }
    ]; };
    /** @nocollapse */ FacilityGroupService.ngInjectableDef = i0.defineInjectable({ factory: function FacilityGroupService_Factory() { return new FacilityGroupService(i0.inject(i1.LocationSelectorService)); }, token: FacilityGroupService, providedIn: "root" });
    return FacilityGroupService;
}());
export { FacilityGroupService };
if (false) {
    /**
     * things that need to emit before generating the facility lanes
     * @type {?}
     */
    FacilityGroupService.prototype.dependency;
    /** @type {?} */
    FacilityGroupService.prototype.locationSelectorService;
}
