// ng-filters-library/esm5/lib/group/group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject, combineLatest, Subject } from 'rxjs';
import { WarningCategoryGroupService } from './warning-category-group.service';
import { HospitalLosGroupService } from './hospital-los-group.service';
import { ProviderGroupService } from './provider-group.service';
import { UnitGroupService } from './unit-group.service';
import { FacilityGroupService } from './facility-group.service';
import * as i0 from "@angular/core";
import * as i1 from "./warning-category-group.service";
import * as i2 from "./hospital-los-group.service";
import * as i3 from "./provider-group.service";
import * as i4 from "./facility-group.service";
import * as i5 from "./unit-group.service";
var GroupService = /** @class */ (function () {
    function GroupService(warningGroupService, hospitalLosGroupService, providerGroupService, facilityGroupService, unitGroupService) {
        var _this = this;
        this.warningGroupService = warningGroupService;
        this.hospitalLosGroupService = hospitalLosGroupService;
        this.providerGroupService = providerGroupService;
        this.facilityGroupService = facilityGroupService;
        this.unitGroupService = unitGroupService;
        /**
         * subject used to change how many groups to be open initially after applying filtering, grouping and/or sorting
         */
        this.changeNumberOfGroupsToShow$ = new BehaviorSubject(4);
        /**
         * subject to expand groups if necessary
         */
        this.expandGroups$ = new BehaviorSubject(false);
        /**
         * switch to decide if it is necessary to recalculate how many groups should be closed
         */
        this.changeGroupsClosed = false;
        /**
         * emits when it is okay to display group options
         */
        this.groupsLoaded$ = new Subject();
        /**
         * different options to group by and corresponding logic to divide graphs into groups
         */
        this.groupOptions = [];
        /** @type {?} */
        var groupOptions = [warningGroupService, hospitalLosGroupService, providerGroupService, facilityGroupService, unitGroupService];
        /** @type {?} */
        var dependencies = this.compileDependencies(groupOptions);
        // set default value of group by
        this.setDefaults();
        // wait for dependencies to load and then
        combineLatest(dependencies).subscribe(function () { return _this.generateGroupByOptions(groupOptions); });
    }
    /**
     * compiles a list of dependencies from all the Group Option classes
     * @param options - list of Services to compile dependencies from
     */
    /**
     * compiles a list of dependencies from all the Group Option classes
     * @param {?} options - list of Services to compile dependencies from
     * @return {?}
     */
    GroupService.prototype.compileDependencies = /**
     * compiles a list of dependencies from all the Group Option classes
     * @param {?} options - list of Services to compile dependencies from
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var dependencies = [];
        try {
            for (var options_1 = tslib_1.__values(options), options_1_1 = options_1.next(); !options_1_1.done; options_1_1 = options_1.next()) {
                var option = options_1_1.value;
                if (option.dependency && dependencies.indexOf(option.dependency) < 0) {
                    dependencies.push(option.dependency);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (options_1_1 && !options_1_1.done && (_a = options_1.return)) _a.call(options_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return dependencies;
        var e_1, _a;
    };
    /**
     * generate filter functions and titles for all the group by options and add them to the options list
     */
    /**
     * generate filter functions and titles for all the group by options and add them to the options list
     * @param {?} options
     * @return {?}
     */
    GroupService.prototype.generateGroupByOptions = /**
     * generate filter functions and titles for all the group by options and add them to the options list
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.groupOptions = [];
        try {
            for (var options_2 = tslib_1.__values(options), options_2_1 = options_2.next(); !options_2_1.done; options_2_1 = options_2.next()) {
                var option = options_2_1.value;
                this.groupOptions.push(option.generateLanes());
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (options_2_1 && !options_2_1.done && (_a = options_2.return)) _a.call(options_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        this.groupsLoaded$.next();
        var e_2, _a;
    };
    /**
     * set default value of group by
     */
    /**
     * set default value of group by
     * @return {?}
     */
    GroupService.prototype.setDefaults = /**
     * set default value of group by
     * @return {?}
     */
    function () {
        this.activeGroup = this.warningGroupService.generateLanes();
        this.stagedGroup = this.warningGroupService.generateLanes();
        this.groupSubject$ = new BehaviorSubject(this.warningGroupService.generateLanes().lanes);
    };
    /**
     * what to do if sidebar is destroyed
     */
    /**
     * what to do if sidebar is destroyed
     * @return {?}
     */
    GroupService.prototype.clearGroupBy = /**
     * what to do if sidebar is destroyed
     * @return {?}
     */
    function () {
        this.activeGroup = this.warningGroupService.generateLanes();
        this.stagedGroup = this.warningGroupService.generateLanes();
        this.groupSubject$.next(this.warningGroupService.generateLanes().lanes);
    };
    GroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    GroupService.ctorParameters = function () { return [
        { type: WarningCategoryGroupService },
        { type: HospitalLosGroupService },
        { type: ProviderGroupService },
        { type: FacilityGroupService },
        { type: UnitGroupService }
    ]; };
    /** @nocollapse */ GroupService.ngInjectableDef = i0.defineInjectable({ factory: function GroupService_Factory() { return new GroupService(i0.inject(i1.WarningCategoryGroupService), i0.inject(i2.HospitalLosGroupService), i0.inject(i3.ProviderGroupService), i0.inject(i4.FacilityGroupService), i0.inject(i5.UnitGroupService)); }, token: GroupService, providedIn: "root" });
    return GroupService;
}());
export { GroupService };
if (false) {
    /**
     * subject used to change how many groups to be open initially after applying filtering, grouping and/or sorting
     * @type {?}
     */
    GroupService.prototype.changeNumberOfGroupsToShow$;
    /**
     * subject to expand groups if necessary
     * @type {?}
     */
    GroupService.prototype.expandGroups$;
    /**
     * switch to decide if it is necessary to recalculate how many groups should be closed
     * @type {?}
     */
    GroupService.prototype.changeGroupsClosed;
    /**
     * emits when it is okay to display group options
     * @type {?}
     */
    GroupService.prototype.groupsLoaded$;
    /**
     * different options to group by and corresponding logic to divide graphs into groups
     * @type {?}
     */
    GroupService.prototype.groupOptions;
    /**
     * tells the groups to update themselves
     * @type {?}
     */
    GroupService.prototype.groupSubject$;
    /**
     * groups active in the array view
     * @type {?}
     */
    GroupService.prototype.activeGroup;
    /**
     * groups currently selected in sidebar
     * @type {?}
     */
    GroupService.prototype.stagedGroup;
    /** @type {?} */
    GroupService.prototype.warningGroupService;
    /** @type {?} */
    GroupService.prototype.hospitalLosGroupService;
    /** @type {?} */
    GroupService.prototype.providerGroupService;
    /** @type {?} */
    GroupService.prototype.facilityGroupService;
    /** @type {?} */
    GroupService.prototype.unitGroupService;
}
