// ng-filters-library/esm5/lib/group/hospital-los-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var HospitalLosGroupService = /** @class */ (function () {
    function HospitalLosGroupService() {
        /**
         * size of LOS lane ranges
         */
        this.losIncrement = 1;
        /**
         * number of lanes LOS group by can have
         */
        this.losLaneMax = 5;
        /**
         * things that need to emit before generating the Hospital Los lanes
         */
        this.dependency = null;
    }
    /**
     * generate group choice object for facilities
     */
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    HospitalLosGroupService.prototype.generateLanes = /**
     * generate group choice object for facilities
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {
            groupName: 'Hospital LOS',
            lanes: []
        };
        for (var i = 0; i < this.losLaneMax * this.losIncrement; i += this.losIncrement) {
            /** @type {?} */
            var losLaneInfo = this.generateLosLaneFilterFunctionAndTitle(i, i + this.losIncrement);
            option.lanes.push({ laneFunction: losLaneInfo.filterFunction, laneTitle: losLaneInfo.title });
        }
        /** @type {?} */
        var losLaneInfoFinal = this.generateLosLaneFilterFunctionAndTitle(this.losLaneMax * this.losIncrement);
        option.lanes.push({ laneFunction: losLaneInfoFinal.filterFunction, laneTitle: losLaneInfoFinal.title });
        return option;
    };
    /**
     * generate filter function and label for Hospital LOS group by
     * @param {?} minIndex - number of days before current time to start including graphs in this lane (inclusive)
     * @param {?=} maxIndex - number of days before current time to stop including graphs (exclusive)
     * @return {?}
     */
    HospitalLosGroupService.prototype.generateLosLaneFilterFunctionAndTitle = /**
     * generate filter function and label for Hospital LOS group by
     * @param {?} minIndex - number of days before current time to start including graphs in this lane (inclusive)
     * @param {?=} maxIndex - number of days before current time to stop including graphs (exclusive)
     * @return {?}
     */
    function (minIndex, maxIndex) {
        if (maxIndex === void 0) { maxIndex = -1; }
        /** @type {?} */
        var minVolume = minIndex * 86400000;
        if (maxIndex !== -1) {
            /** @type {?} */
            var maxVolume_1 = maxIndex * 86400000;
            return {
                filterFunction: function (peraGraph) {
                    /** @type {?} */
                    var current = ((new Date()).getTime());
                    /** @type {?} */
                    var min = current - minVolume;
                    /** @type {?} */
                    var max = current - maxVolume_1;
                    return peraGraph.admitDate <= min && peraGraph.admitDate > max;
                }, title: minIndex.toString() + ' - ' + maxIndex.toString() + ' days'
            };
        }
        else {
            // create catch all if max is not included in function call
            return {
                filterFunction: function (peraGraph) {
                    /** @type {?} */
                    var current = ((new Date()).getTime());
                    /** @type {?} */
                    var min = current - minVolume;
                    return peraGraph.admitDate <= min;
                }, title: minIndex.toString() + '+ days'
            };
        }
    };
    HospitalLosGroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    HospitalLosGroupService.ctorParameters = function () { return []; };
    /** @nocollapse */ HospitalLosGroupService.ngInjectableDef = i0.defineInjectable({ factory: function HospitalLosGroupService_Factory() { return new HospitalLosGroupService(); }, token: HospitalLosGroupService, providedIn: "root" });
    return HospitalLosGroupService;
}());
export { HospitalLosGroupService };
if (false) {
    /**
     * size of LOS lane ranges
     * @type {?}
     */
    HospitalLosGroupService.prototype.losIncrement;
    /**
     * number of lanes LOS group by can have
     * @type {?}
     */
    HospitalLosGroupService.prototype.losLaneMax;
    /**
     * things that need to emit before generating the Hospital Los lanes
     * @type {?}
     */
    HospitalLosGroupService.prototype.dependency;
}
