// ng-filters-library/esm5/lib/group/provider-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ProviderFilterService } from '../filter/providers/provider-filter.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../filter/providers/provider-filter.service";
var ProviderGroupService = /** @class */ (function () {
    function ProviderGroupService(providerService) {
        var _this = this;
        this.providerService = providerService;
        /**
         * things that need to emit before generating the provider lanes
         */
        this.dependency = new Subject();
        this.providerService.providersLoaded.subscribe(function () {
            _this.dependency.next();
        });
    }
    /**
     * generate group choice object for facilities
     */
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    ProviderGroupService.prototype.generateLanes = /**
     * generate group choice object for facilities
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {
            groupName: 'Provider',
            lanes: []
        };
        try {
            for (var _a = tslib_1.__values(this.providerService.providers), _b = _a.next(); !_b.done; _b = _a.next()) {
                var provider = _b.value;
                option.lanes.push({
                    laneFunction: this.generateProviderLaneFilterFunction(provider),
                    laneTitle: provider.lastName + ', ' + provider.firstName
                });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return option;
        var e_1, _c;
    };
    /**
     * helper function for generating a filter function for provider groups given a provider
     * @param {?} provider - provider to filter for
     * @return {?}
     */
    ProviderGroupService.prototype.generateProviderLaneFilterFunction = /**
     * helper function for generating a filter function for provider groups given a provider
     * @param {?} provider - provider to filter for
     * @return {?}
     */
    function (provider) {
        /** @type {?} */
        var firstName = provider.firstName;
        /** @type {?} */
        var lastName = provider.lastName;
        return function (peraGraph) {
            return peraGraph.attendingProvider.firstName === firstName && peraGraph.attendingProvider.lastName === lastName;
        };
    };
    ProviderGroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ProviderGroupService.ctorParameters = function () { return [
        { type: ProviderFilterService }
    ]; };
    /** @nocollapse */ ProviderGroupService.ngInjectableDef = i0.defineInjectable({ factory: function ProviderGroupService_Factory() { return new ProviderGroupService(i0.inject(i1.ProviderFilterService)); }, token: ProviderGroupService, providedIn: "root" });
    return ProviderGroupService;
}());
export { ProviderGroupService };
if (false) {
    /**
     * things that need to emit before generating the provider lanes
     * @type {?}
     */
    ProviderGroupService.prototype.dependency;
    /** @type {?} */
    ProviderGroupService.prototype.providerService;
}
