// ng-filters-library/esm5/lib/group/unit-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { LocationSelectorService } from '../filter/locations/location-selector.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../filter/locations/location-selector.service";
var UnitGroupService = /** @class */ (function () {
    function UnitGroupService(locationSelectorService) {
        var _this = this;
        this.locationSelectorService = locationSelectorService;
        /**
         * things that need to emit before generating the unit lanes
         */
        this.dependency = new Subject();
        this.locationSelectorService.loadedLocations.subscribe(function () {
            _this.dependency.next();
        });
    }
    /**
     * generate group choice object for facilities
     */
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    UnitGroupService.prototype.generateLanes = /**
     * generate group choice object for facilities
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {
            groupName: 'Unit',
            lanes: []
        };
        try {
            for (var _a = tslib_1.__values(this.locationSelectorService.units), _b = _a.next(); !_b.done; _b = _a.next()) {
                var unit = _b.value;
                option.lanes.push({ laneFunction: this.generateUnitLaneFilterFunction(unit), laneTitle: unit.displayName });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return option;
        var e_1, _c;
    };
    /**
     * generate filter function for filtering out a group for a specific unit
     * @param {?} unit - unit to filter on
     * @return {?}
     */
    UnitGroupService.prototype.generateUnitLaneFilterFunction = /**
     * generate filter function for filtering out a group for a specific unit
     * @param {?} unit - unit to filter on
     * @return {?}
     */
    function (unit) {
        /** @type {?} */
        var unitName = unit.displayName;
        return function (peraGraph) {
            return peraGraph.unitId === unitName;
        };
    };
    UnitGroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    UnitGroupService.ctorParameters = function () { return [
        { type: LocationSelectorService }
    ]; };
    /** @nocollapse */ UnitGroupService.ngInjectableDef = i0.defineInjectable({ factory: function UnitGroupService_Factory() { return new UnitGroupService(i0.inject(i1.LocationSelectorService)); }, token: UnitGroupService, providedIn: "root" });
    return UnitGroupService;
}());
export { UnitGroupService };
if (false) {
    /**
     * things that need to emit before generating the unit lanes
     * @type {?}
     */
    UnitGroupService.prototype.dependency;
    /** @type {?} */
    UnitGroupService.prototype.locationSelectorService;
}
