// ng-filters-library/esm5/lib/group/warning-category-group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var WarningCategoryGroupService = /** @class */ (function () {
    function WarningCategoryGroupService() {
        /**
         * things that need to emit before generating the warning lanes
         */
        this.dependency = null;
    }
    /**
     * generate group choice object for facilities
     */
    /**
     * generate group choice object for facilities
     * @return {?}
     */
    WarningCategoryGroupService.prototype.generateLanes = /**
     * generate group choice object for facilities
     * @return {?}
     */
    function () {
        return {
            groupName: 'Warning Category',
            lanes: [{
                    laneFunction: function (peraGraph) { return peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'V'; },
                    laneTitle: 'Very High Warning'
                },
                {
                    laneFunction: function (peraGraph) { return peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'H'; },
                    laneTitle: 'High Warning'
                },
                {
                    laneFunction: function (peraGraph) { return peraGraph.activeWarning != null && peraGraph.activeWarning.category === 'M'; },
                    laneTitle: 'Medium Warning'
                },
                { laneFunction: function (peraGraph) { return peraGraph.activeWarning == null; }, laneTitle: 'No Warning' }]
        };
    };
    WarningCategoryGroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    WarningCategoryGroupService.ctorParameters = function () { return []; };
    /** @nocollapse */ WarningCategoryGroupService.ngInjectableDef = i0.defineInjectable({ factory: function WarningCategoryGroupService_Factory() { return new WarningCategoryGroupService(); }, token: WarningCategoryGroupService, providedIn: "root" });
    return WarningCategoryGroupService;
}());
export { WarningCategoryGroupService };
if (false) {
    /**
     * things that need to emit before generating the warning lanes
     * @type {?}
     */
    WarningCategoryGroupService.prototype.dependency;
}
