// ng-login-library/esm5/lib/login.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LoginComponent } from './login.component';
import { LoginService } from './services/login.service';
import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { LoginConfigService } from './services/login-config.service';
import { NgModule } from '@angular/core';
import { LoginIpService } from './services/login-ip.service';
import { HttpClientModule } from '@angular/common/http';
var LoginModule = /** @class */ (function () {
    function LoginModule() {
    }
    /**
     * @return {?}
     */
    LoginModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: LoginModule,
            providers: [
                HttpClientModule,
                LoginService,
                LoginIpService,
                LoginConfigService
            ]
        };
    };
    LoginModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        BrowserModule,
                        FormsModule
                    ],
                    declarations: [LoginComponent],
                    exports: [LoginComponent]
                },] }
    ];
    return LoginModule;
}());
export { LoginModule };
