// ng-login-library/esm5/lib/login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChildren, ViewEncapsulation } from '@angular/core';
import { LoginService } from './services/login.service';
import { LoginIp } from './model/LoginIp';
import { LoginIpService } from './services/login-ip.service';
import { User } from 'ng-common-library';
import { first } from 'rxjs/operators';
var LoginComponent = /** @class */ (function () {
    function LoginComponent(loginService, loginLibraryIpService) {
        this.loginService = loginService;
        this.loginLibraryIpService = loginLibraryIpService;
        this.NO_DICE = 'Error:  Username or password is incorrect';
        this.loginInfo = new LoginIp();
    }
    /**
     * @return {?}
     */
    LoginComponent.prototype.getUser = /**
     * @return {?}
     */
    function () {
        if (!this.loginService.getCurrentUser()) {
            this.loginService.setCurrentUser(new User());
        }
        return this.loginService.getCurrentUser();
    };
    /**
     * @return {?}
     */
    LoginComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.logoSrc = this.loginService.getLogoSrc();
        this.loginLibraryIpService.loginIp$.pipe(first()).subscribe(function (loginIp) {
            if (loginIp) {
                _this.loginInfo = loginIp;
            }
            else {
                console.log('Could not get Login Information from server, using defaults.');
            }
        });
        this.loginLibraryIpService.getLoginIp();
        localStorage.removeItem('currentUser');
        this.loginService.setCurrentUser(new User());
        this.errorMessage = null;
    };
    /**
     * @return {?}
     */
    LoginComponent.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Retrieve a reference to the current user from the loginService
        /** @type {?} */
        var user = this.getUser();
        if (user.username && this.userPassword) {
            // Base64 encode the password
            user.password = btoa(this.userPassword);
            this.loginService.login$.pipe(first()).subscribe(function (usr) {
                if (!usr) {
                    /*
                       Any error = this message. This is a security best practice as you
                       do not want to give the user any information that they can use to
                       get into the system.
                       Wipe out the encoded password
                     */
                    user.password = null;
                    _this.userPassword = null;
                    _this.errorMessage = _this.NO_DICE;
                    _this.loginErrored();
                }
                else {
                    // Wipe out the encoded password
                    user.password = null;
                    _this.errorMessage = null;
                    _this.userPassword = null;
                }
            });
            this.loginService.logIn(user);
        }
        else {
            this.errorMessage = this.NO_DICE;
            this.loginErrored();
        }
    };
    /**
     * @return {?}
     */
    LoginComponent.prototype.loginErrored = /**
     * @return {?}
     */
    function () {
        this.setFocusToUsernameField();
    };
    /**
     * @return {?}
     */
    LoginComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.setFocusToUsernameField();
    };
    /**
     * @return {?}
     */
    LoginComponent.prototype.setFocusToUsernameField = /**
     * @return {?}
     */
    function () {
        if (this.vc && this.vc.first && this.vc.first.nativeElement) {
            this.vc.first.nativeElement.focus();
        }
    };
    LoginComponent.decorators = [
        { type: Component, args: [{
                    selector: 'phlib-login',
                    template: "<!-- Hide the component if the user is already logged in. -->\n<div class=\"container\" [hidden]=\"loginService.hideLogin\">\n  <div class=\"card card-container\">\n    <img id=\"profile-img\" class=\"profile-img-card\" src={{logoSrc}} />\n    <p id=\"login-subtitle\" class=\"login-subtitle\">\n        Log Into Your <strong>{{title}}</strong> Account\n    </p>\n\n    <form name=\"form-signin\" (ngSubmit)=\"f.form.valid && login()\" #f=\"ngForm\" novalidate>\n      <div class=\"alert alert-danger\" role=\"alert\" *ngIf=\"errorMessage\">{{errorMessage}}</div>\n      <span id=\"reauth-email\" class=\"reauth-email\"></span>\n      <div class=\"form-group\" [ngClass]=\"{ 'has-error': f.submitted && !username.valid }\">\n        <label for=\"username\">User Name</label>\n        <input #input autofocus type=\"text\" class=\"form-control\" id=\"username\" name=\"username\" [(ngModel)]=\"getUser().username\" #username=\"ngModel\" required />\n      </div>\n      <div class=\"form-group\" [ngClass]=\"{ 'has-error': f.submitted && !password.valid }\">\n        <label for=\"password\">Password</label>\n        <input type=\"password\" class=\"form-control\" id=\"password\" name=\"password\" autocomplete=\"off\" [(ngModel)]=\"userPassword\" #password=\"ngModel\" required />\n      </div>\n      <div id=\"remember\" class=\"checkbox\">\n\n      </div>\n      <button [disabled]=\"username.invalid || password.invalid\" class=\"btn btn-lg btn-primary btn-block btn-signin\" type=\"submit\">LOGIN</button>\n    </form>\n  </div>\n  <footer>\n    <div class=\"footer\">\n      <span>{{loginInfo.copyRightInfo}} <strong>Version:</strong> {{loginInfo.versionInfo}}\n          | <strong>UDI: </strong>{{loginInfo.uniqueDeviceIdentifier}} <span [innerHtml]=\"loginInfo.patentInfo\"></span>\n      </span>\n    </div>\n  </footer>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["body,html{height:100%;background:no-repeat #3b3b3b}.ph-login-button{display:none}.card-container.card{max-width:490px;height:500px;padding:33px 55px 0}.btn{font-weight:700;height:36px;-moz-user-select:none;-webkit-user-select:none;-ms-user-select:none;user-select:none;cursor:default}.card{background-color:#fff;padding:20px 25px 30px;margin:100px auto 25px;border-radius:4px;box-shadow:0 2px 2px rgba(0,0,0,.3)}.profile-img-card{width:100%;height:3.5em;margin:0 auto 10px;display:block}.login-subtitle{font-size:1.2em;font-weight:700;color:#999;font-family:Tahoma;margin:10px 0 20px;min-height:1em}.reauth-email{display:block;color:#404040;line-height:2;margin-bottom:10px;font-size:14px;text-align:center;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;box-sizing:border-box}.form-signin #inputEmail,.form-signin #inputPassword{direction:ltr;height:44px;font-size:16px}label{color:#d9d9d9;text-transform:uppercase;font-weight:400;font-size:14px;font-family:Tahoma}.form-signin button,.form-signin input[type=email],.form-signin input[type=password],.form-signin input[type=text]{width:100%;display:block;margin-bottom:10px;z-index:1;position:relative;box-sizing:border-box}.form-control:focus{border-left-color:#fff;border-right-color:#fff;border-top-color:#fff;box-shadow:none;-webkit-box-shadow:none}.form-control{background-color:#fff;border-radius:0}.form-group input{border-top:none;border-left:none;border-right:none;box-shadow:none;margin-bottom:2em;font-size:1.2em}.btn.btn-signin{background-color:#999;display:block;margin-right:auto;margin-left:auto;padding:0;font-weight:700;font-size:14px;font-family:Tahoma;width:50%;height:36px;border-radius:3px;border:#666;transition:218ms}.forgot-password{color:#6891a2}.forgot-password:active,.forgot-password:focus,.forgot-password:hover{color:#0c6121}.footer{font-size:.7em;position:fixed;width:100%;bottom:0;left:0;padding-left:5px;background-color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    LoginComponent.ctorParameters = function () { return [
        { type: LoginService },
        { type: LoginIpService }
    ]; };
    LoginComponent.propDecorators = {
        vc: [{ type: ViewChildren, args: ['input',] }]
    };
    return LoginComponent;
}());
export { LoginComponent };
if (false) {
    /** @type {?} */
    LoginComponent.prototype.vc;
    /**
     * @type {?}
     * @private
     */
    LoginComponent.prototype.NO_DICE;
    /** @type {?} */
    LoginComponent.prototype.loginInfo;
    /** @type {?} */
    LoginComponent.prototype.errorMessage;
    /** @type {?} */
    LoginComponent.prototype.logoSrc;
    /** @type {?} */
    LoginComponent.prototype.title;
    /** @type {?} */
    LoginComponent.prototype.userPassword;
    /** @type {?} */
    LoginComponent.prototype.loginService;
    /** @type {?} */
    LoginComponent.prototype.loginLibraryIpService;
}
