// ng-common-library/esm2015/lib/service/token/config/token-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class TokenConfigService {
    constructor() {
        this.checkTokenEndPoint = null;
        this.refreshTokenEndPoint = null;
        this.getTokenEndPoint = null;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setCheckTokenEndPoint(value) {
        this.checkTokenEndPoint = value;
    }
    /**
     * @return {?}
     */
    getCheckTokenEndPoint() {
        if (!this.checkTokenEndPoint) {
            console.warn('Check token endpoint is not set!');
        }
        return this.checkTokenEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setRefreshTokenEndPoint(value) {
        this.refreshTokenEndPoint = value;
    }
    /**
     * @return {?}
     */
    getRefreshTokenEndPoint() {
        if (!this.refreshTokenEndPoint) {
            console.warn('Refresh token endpoint is not set!');
        }
        return this.refreshTokenEndPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setGetTokenEndPoint(value) {
        this.getTokenEndPoint = value;
    }
    /**
     * @return {?}
     */
    getGetTokenEndPoint() {
        if (!this.getTokenEndPoint) {
            console.warn('Get token endpoint is not set!');
        }
        return this.getTokenEndPoint;
    }
}
TokenConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ TokenConfigService.ngInjectableDef = i0.defineInjectable({ factory: function TokenConfigService_Factory() { return new TokenConfigService(); }, token: TokenConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    TokenConfigService.prototype.checkTokenEndPoint;
    /** @type {?} */
    TokenConfigService.prototype.refreshTokenEndPoint;
    /** @type {?} */
    TokenConfigService.prototype.getTokenEndPoint;
}
