// ng-filters-library/esm5/lib/filter/warnings/warnings-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var WarningsFilterService = /** @class */ (function () {
    function WarningsFilterService() {
        /**
         * Warning levels that are selected. Start with all of them.
         */
        this.selectedWarnings = ['warning_v', 'warning_h', 'warning_m', 'warning_none', 'warning_no_ri'];
        /**
         * Configurations for warning level filter options.
         *
         * Contains a filter id, the color of the warning level and a filter function for filtering out
         * the warning if NOT selected.
         */
        this.filterOptions = [
            {
                filterID: 'warning_v', color: '#D51C19',
                filterFunction: function (peragraph) {
                    return !(peragraph.activeWarning && peragraph.activeWarning.category === 'V');
                }
            },
            {
                filterID: 'warning_h', color: '#EE834C',
                filterFunction: function (peragraph) {
                    return !(peragraph.activeWarning && peragraph.activeWarning.category === 'H');
                }
            },
            {
                filterID: 'warning_m', color: '#F6CC5D',
                filterFunction: function (peragraph) {
                    return !(peragraph.activeWarning && peragraph.activeWarning.category === 'M');
                }
            },
            {
                filterID: 'warning_none', color: '#3185B8',
                filterFunction: function (peragraph) {
                    return !(!peragraph.activeWarning && !peragraph.discharged);
                }
            }
        ];
    }
    WarningsFilterService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ WarningsFilterService.ngInjectableDef = i0.defineInjectable({ factory: function WarningsFilterService_Factory() { return new WarningsFilterService(); }, token: WarningsFilterService, providedIn: "root" });
    return WarningsFilterService;
}());
export { WarningsFilterService };
if (false) {
    /**
     * Warning levels that are selected. Start with all of them.
     * @type {?}
     */
    WarningsFilterService.prototype.selectedWarnings;
    /**
     * Configurations for warning level filter options.
     *
     * Contains a filter id, the color of the warning level and a filter function for filtering out
     * the warning if NOT selected.
     * @type {?}
     */
    WarningsFilterService.prototype.filterOptions;
}
