// ng-common-library/esm5/lib/service/diagnosis/config/diagnosis-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var DiagnosisConfigService = /** @class */ (function () {
    function DiagnosisConfigService() {
        this.primaryDiagnosesEndpoint = null;
    }
    /**
     * @return {?}
     */
    DiagnosisConfigService.prototype.getPrimaryDiagnosesEndpoint = /**
     * @return {?}
     */
    function () {
        if (!this.primaryDiagnosesEndpoint) {
            console.warn('Primary diagnoses endpoint is not set!');
        }
        return this.primaryDiagnosesEndpoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DiagnosisConfigService.prototype.setPrimaryDiagnosesEndpoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.primaryDiagnosesEndpoint = value;
    };
    DiagnosisConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ DiagnosisConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DiagnosisConfigService_Factory() { return new DiagnosisConfigService(); }, token: DiagnosisConfigService, providedIn: "root" });
    return DiagnosisConfigService;
}());
export { DiagnosisConfigService };
if (false) {
    /** @type {?} */
    DiagnosisConfigService.prototype.primaryDiagnosesEndpoint;
}
