// ng-common-library/esm5/lib/service/flowsheets/api/flowsheets-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { FlowsheetsConfigService } from '../config/flowsheets-config.service';
import { BehaviorSubject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/flowsheets-config.service";
var FlowsheetsApiService = /** @class */ (function () {
    function FlowsheetsApiService(http, flowsheetConfigService) {
        this.http = http;
        this.flowsheetConfigService = flowsheetConfigService;
        this.pollInstance$ = timer(0, 60000);
        this.flowSheetSubject = new BehaviorSubject(null);
        this.riScoresSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getFlowSheetSubject = /**
     * @return {?}
     */
    function () {
        return this.flowSheetSubject;
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getFlowSheetSubjectValue = /**
     * @return {?}
     */
    function () {
        return this.flowSheetSubject.value;
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getRiScoresSubject = /**
     * @return {?}
     */
    function () {
        return this.riScoresSubject;
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getRiScoresSubjectValue = /**
     * @return {?}
     */
    function () {
        return this.riScoresSubject.value;
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getSinglePeraGraphFlowSheet = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.flowsheetConfigService.getSinglePeraGraphFlowSheetEndPoint());
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getDetailedFlowsheetColumn = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.flowsheetConfigService.getDetailedFlowSheetColumnEndpoint());
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getAllPeraGraphFlowSheets = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.flowsheetConfigService.getAllPeragraphFlowSheetsEndPoint(), { observe: 'response' });
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getAllRiScores = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.flowsheetConfigService.getLatestRiScoresEndPoint(), { observe: 'response' });
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getFlowSheetInterval = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(function () { return _this.getAllPeraGraphFlowSheets(); }));
    };
    /**
     * @return {?}
     */
    FlowsheetsApiService.prototype.getLatestRiScoresInterval = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Make an HTTP call on an interval
        return this.pollInstance$.pipe(switchMap(function () { return _this.getAllRiScores(); }));
    };
    FlowsheetsApiService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FlowsheetsApiService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: FlowsheetsConfigService }
    ]; };
    /** @nocollapse */ FlowsheetsApiService.ngInjectableDef = i0.defineInjectable({ factory: function FlowsheetsApiService_Factory() { return new FlowsheetsApiService(i0.inject(i1.HttpClient), i0.inject(i2.FlowsheetsConfigService)); }, token: FlowsheetsApiService, providedIn: "root" });
    return FlowsheetsApiService;
}());
export { FlowsheetsApiService };
if (false) {
    /** @type {?} */
    FlowsheetsApiService.prototype.flowSheetSubject;
    /** @type {?} */
    FlowsheetsApiService.prototype.riScoresSubject;
    /** @type {?} */
    FlowsheetsApiService.prototype.pollInstance$;
    /** @type {?} */
    FlowsheetsApiService.prototype.http;
    /** @type {?} */
    FlowsheetsApiService.prototype.flowsheetConfigService;
}
