// ng-array-view-library/esm5/lib/service/array-view-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as moment_ from 'moment';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
// The JavaScript compiler gives a weird "moment cannot be used as a namespace" type of error
// This is the workaround...
/** @type {?} */
var moment = moment_;
var ArrayViewConfigService = /** @class */ (function () {
    function ArrayViewConfigService() {
        this.configurationLoaded$ = new BehaviorSubject(false);
        this.currentTimestamp = null;
        this.toolTipInfoEndPoint = null;
    }
    /**
     * @return {?}
     */
    ArrayViewConfigService.prototype.getToolTipInfoEndPoint = /**
     * @return {?}
     */
    function () {
        if (!this.toolTipInfoEndPoint) {
            console.warn('Tooltip endpoint is not set!');
        }
        return this.toolTipInfoEndPoint;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    ArrayViewConfigService.prototype.setToolTipInfoEndPoint = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        this.toolTipInfoEndPoint = val;
    };
    /**
     * @return {?}
     */
    ArrayViewConfigService.prototype.getCurrentTimestamp = /**
     * @return {?}
     */
    function () {
        return this.currentTimestamp;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ArrayViewConfigService.prototype.setCurrentTimestamp = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.currentTimestamp = value;
    };
    ArrayViewConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ArrayViewConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ArrayViewConfigService_Factory() { return new ArrayViewConfigService(); }, token: ArrayViewConfigService, providedIn: "root" });
    return ArrayViewConfigService;
}());
export { ArrayViewConfigService };
if (false) {
    /** @type {?} */
    ArrayViewConfigService.prototype.configurationLoaded$;
    /**
     * @type {?}
     * @private
     */
    ArrayViewConfigService.prototype.currentTimestamp;
    /**
     * @type {?}
     * @private
     */
    ArrayViewConfigService.prototype.toolTipInfoEndPoint;
}
